// @kastov/cryptohapp/build/backend/crypto/create-happ-crypto-link.backend.js
"use strict";
// Reference: https://www.happ.su/main/developer-documentation/crypto-link#api-instructions
Object.defineProperty(exports, "__esModule", { value: true });
exports.createHappCryptoLink = createHappCryptoLink;
const node_crypto_1 = require("node:crypto");
const constants_1 = require("../constants");
const RSA_PKCS1_PADDING = 1;
const CRYPTO_CONFIGS = {
    v2: constants_1.HAPP_CRYPTO_V2,
    v3: constants_1.HAPP_CRYPTO_V3,
    v4: constants_1.HAPP_CRYPTO_V4,
};
function createHappCryptoLink(content, version, asLink) {
    try {
        const config = CRYPTO_CONFIGS[version];
        const options = {
            key: config.publicKey,
            padding: RSA_PKCS1_PADDING,
        };
        const encrypted = (0, node_crypto_1.publicEncrypt)(options, Buffer.from(content));
        const encryptedContent = encrypted.toString('base64');
        if (asLink) {
            return config.deepLink + encryptedContent;
        }
        return {
            deepLink: config.deepLink,
            encryptedContent,
        };
    }
    catch {
        return null;
    }
}
