// @remnawave/backend-contract/build/frontend/models/response-rules/response-rule-modifications.schema.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ResponseRuleModificationsSchema = void 0;
const zod_1 = __importDefault(require("zod"));
exports.ResponseRuleModificationsSchema = zod_1.default
    .object({
    headers: zod_1.default
        .array(zod_1.default
        .object({
        key: zod_1.default
            .string()
            .regex(/^[!#$%&'*+\-.0-9A-Z^_`a-z|~]+$/, 'Invalid header name. Only letters(a-z, A-Z), numbers(0-9), underscores(_) and hyphens(-) are allowed.')
            .describe(JSON.stringify({
            markdownDescription: 'Key of the response header. Must comply with RFC 7230.',
        })),
        value: zod_1.default
            .string()
            .min(1, 'Value is required')
            .describe(JSON.stringify({
            markdownDescription: 'Value of the response header. ',
        })),
    })
        .describe(JSON.stringify({
        markdownDescription: '**Key** and **value** of the response header will be added to the response.',
    })))
        .describe(JSON.stringify({
        defaultSnippets: [
            {
                label: 'Examples: Add custom header',
                markdownDescription: 'Add a custom header to the response',
                body: [
                    {
                        key: 'X-Custom-Header',
                        value: 'CustomValue',
                    },
                ],
            },
        ],
        markdownDescription: 'Array of headers to be added when the rule is matched.',
    }))
        .optional(),
    subscriptionTemplate: zod_1.default
        .string()
        .min(1, 'Subscription template name is required')
        .optional()
        .describe(JSON.stringify({
        markdownDescription: 'Override the subscription template with the given name. If not provided, the default subscription template will be used. If the template name is not found, the default subscription template for this type will be used. **This modification have higher priority than settings from External Squads.**',
    })),
})
    .optional()
    .describe(JSON.stringify({
    examples: [
        {
            headers: [
                {
                    key: 'X-Custom-Header',
                    value: 'CustomValue',
                },
            ],
        },
    ],
    markdownDescription: 'Response modifications to be applied when the rule is matched. Optional.',
}));
