// steiger/dist/chunk-BLZHJTX3.mjs
import module from 'module';
import path, { dirname, sep, join, relative, basename, resolve, isAbsolute } from 'path';
import url from 'url';
import { createStore, createEvent, combine, createEffect, sample, merge } from 'effector';
import { debounce, not, combineEvents } from 'patronum';
import chokidar from 'chokidar';
import * as find from 'empathic/find';
import { isGitIgnored } from 'globby';
import { produce } from 'immer';
import z from 'zod';
import { isEqual as isEqual$1 } from 'lodash-es';
import micromatch from 'micromatch';
import { access, stat } from 'fs/promises';

var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __require = /* @__PURE__ */ ((x) => typeof require !== "undefined" ? require : typeof Proxy !== "undefined" ? new Proxy(x, {
  get: (a, b) => (typeof require !== "undefined" ? require : a)[b]
}) : x)(function(x) {
  if (typeof require !== "undefined") return require.apply(this, arguments);
  throw Error('Dynamic require of "' + x + '" is not supported');
});
var __esm = (fn, res) => function __init() {
  return fn && (res = (0, fn[__getOwnPropNames(fn)[0]])(fn = 0)), res;
};
var __commonJS = (cb, mod) => function __require2() {
  return mod || (0, cb[__getOwnPropNames(cb)[0]])((mod = { exports: {} }).exports, mod), mod.exports;
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var init_cjs_shim = __esm({
  "cjs-shim.ts"() {
    globalThis.require = module.createRequire(import.meta.url);
    globalThis.__filename = url.fileURLToPath(import.meta.url);
    globalThis.__dirname = path.dirname(__filename);
  }
});

// src/app.ts
init_cjs_shim();

// src/features/transfer-fs-to-vfs.ts
init_cjs_shim();

// src/models/vfs.ts
init_cjs_shim();
function createVfsRoot(rootPath) {
  const rootFolder = { type: "folder", path: rootPath, children: [] };
  const $tree = createStore(rootFolder);
  const fileAdded = createEvent();
  $tree.on(
    fileAdded,
    (state, newFilePath) => produce(state, (draft) => {
      const pathSegments = relative(rootPath, newFilePath).split(sep);
      let currentFolder = draft;
      for (const pathSegment of pathSegments.slice(0, -1)) {
        const existingChild = currentFolder.children.find(
          (child) => child.type === "folder" && basename(child.path) === pathSegment
        );
        if (existingChild === void 0) {
          currentFolder.children.push({
            type: "folder",
            path: join(currentFolder.path, pathSegment),
            children: []
          });
          currentFolder = currentFolder.children[currentFolder.children.length - 1];
        } else {
          currentFolder = existingChild;
        }
      }
      currentFolder.children.push({ type: "file", path: newFilePath });
    })
  );
  const fileRemoved = createEvent();
  $tree.on(
    fileRemoved,
    (state, removedFilePath) => produce(state, (draft) => {
      const pathSegments = relative(rootPath, removedFilePath).split(sep);
      const parents = [draft];
      let currentFolder = draft;
      for (const pathSegment of pathSegments.slice(0, -1)) {
        const existingChild = currentFolder.children.find(
          (child) => child.type === "folder" && basename(child.path) === pathSegment
        );
        if (existingChild === void 0) {
          throw Error(`Folder ${pathSegment} not found`);
        } else {
          currentFolder = existingChild;
          parents.push(currentFolder);
        }
      }
      const removedFileIndex = currentFolder.children.findIndex(
        (child) => child.type === "file" && child.path === removedFilePath
      );
      if (removedFileIndex === -1) {
        throw Error(`File ${removedFilePath} not found`);
      }
      currentFolder.children.splice(removedFileIndex, 1);
      parents.reverse().forEach((parent, index) => {
        if (parent.children.length === 0 && parents[index + 1] !== void 0) {
          parents[index + 1].children.splice(parents[index + 1].children.indexOf(parent), 1);
        }
      });
    })
  );
  return { $tree, fileAdded, fileRemoved };
}

// src/features/transfer-fs-to-vfs.ts
async function createWatcher(path2) {
  const vfs = createVfsRoot(path2);
  const gitFolder = find.up(".git", { cwd: path2 });
  const isIgnored = await isGitIgnored({ cwd: gitFolder ? dirname(gitFolder) : path2 });
  const watcher = chokidar.watch(path2, {
    ignored: (path3) => path3.split(sep).includes("node_modules") || path3.split(sep).includes(".git") || isIgnored(path3),
    ignoreInitial: false,
    alwaysStat: true,
    awaitWriteFinish: true,
    cwd: path2
  });
  watcher.on("add", async (relativePath) => {
    vfs.fileAdded(join(path2, relativePath));
  });
  watcher.on("unlink", async (relativePath) => {
    vfs.fileRemoved(join(path2, relativePath));
  });
  return {
    vfs,
    watcher
  };
}
function scan(path2) {
  return new Promise(
    (resolve2, reject) => createWatcher(path2).then(({ vfs, watcher }) => {
      watcher.on("ready", () => {
        watcher.close();
        resolve2(vfs.$tree.getState());
      });
    }).catch(reject)
  );
}

// src/shared/defer.ts
init_cjs_shim();
var defer = (args) => {
  const { clock, until: condition } = args;
  const calledAfterCondition = sample({
    clock,
    filter: condition
  });
  const calledBeforeCondition = sample({
    clock,
    filter: not(condition)
  });
  return merge([
    calledAfterCondition,
    combineEvents({
      events: [calledBeforeCondition, condition.updates.filter({ fn: Boolean })],
      reset: condition.updates.filter({ fn: (value) => !value })
    })
  ]);
};

// src/models/config/index.ts
init_cjs_shim();

// src/models/config/create-rule-instructions.ts
init_cjs_shim();

// src/models/config/raw-config.ts
init_cjs_shim();
function getSeverity(severityOrTuple) {
  return Array.isArray(severityOrTuple) ? severityOrTuple[0] : severityOrTuple;
}
function getOptions(severityOrTuple) {
  return Array.isArray(severityOrTuple) ? severityOrTuple[1] : null;
}
function isGlobalIgnore(obj) {
  return typeof obj === "object" && obj !== null && "ignores" in obj && Object.keys(obj).length === 1;
}
function isConfigObject(obj) {
  return typeof obj === "object" && obj !== null && "rules" in obj;
}
function isPlugin(obj) {
  return typeof obj === "object" && obj !== null && "ruleDefinitions" in obj;
}
function isConfiguration(obj) {
  return isConfigObject(obj) || isGlobalIgnore(obj);
}

// src/models/config/create-rule-instructions.ts
function createEmptyInstructions() {
  return {
    options: null,
    globGroups: []
  };
}
function extractRuleNames(configObject) {
  return Object.keys(configObject.rules);
}
function preCreateRuleInstructions(config) {
  return config.filter(isConfigObject).flatMap(extractRuleNames).reduce(
    (acc, item) => ({
      ...acc,
      [item]: createEmptyInstructions()
    }),
    {}
  );
}
function createRuleInstructions(config) {
  const ruleNameToInstructions = preCreateRuleInstructions(config);
  return config.reduce((acc, item) => {
    if (isConfigObject(item)) {
      Object.entries(item.rules).forEach(([ruleName, severityOrTuple]) => {
        if (severityOrTuple === void 0) {
          return;
        }
        const ruleOptions = getOptions(severityOrTuple);
        if (ruleOptions) {
          acc[ruleName].options = ruleOptions;
        }
        acc[ruleName].globGroups.push({
          severity: getSeverity(severityOrTuple),
          files: item.files,
          ignores: item.ignores
        });
      });
    }
    return acc;
  }, ruleNameToInstructions);
}

// src/models/config/validate-config.ts
init_cjs_shim();

// src/shared/objects.ts
init_cjs_shim();
var isEqual = isEqual$1;

// src/models/config/validate-config.ts
var OLD_CONFIG_ERROR_MESSAGE = "Old configuration format detected. We are evolving!\nPlease follow this short guide to migrate to the new one:\nhttps://github.com/feature-sliced/steiger/blob/master/MIGRATION_GUIDE.md";
var WRONG_CONFIG_SHAPE_ERROR_MESSAGE = "The config should be an Array, but the provided config is not.\nHere is a link to the documentation that might help to fix it:\nhttps://github.com/feature-sliced/steiger?tab=readme-ov-file#configuration";
var NO_RULES_ERROR_MESSAGE = "At least one rule must be provided by plugins!";
var NO_CONFIG_OBJECTS_ERROR_MESSAGE = "At least one config object must be provided!";
function getAllRuleNames(plugins) {
  const allRules = plugins.flatMap((plugin) => plugin.ruleDefinitions);
  return allRules.map((rule) => rule.name);
}
function validateConfigObjectsNumber(value, ctx) {
  const configObjects = value.filter(isConfigObject);
  if (configObjects.length === 0) {
    ctx.addIssue({
      code: z.ZodIssueCode.custom,
      message: NO_CONFIG_OBJECTS_ERROR_MESSAGE
    });
  }
}
function validateRuleUniqueness(value, ctx) {
  const allRuleNames = getAllRuleNames(value.filter(isPlugin));
  const uniqueNames = new Set(allRuleNames);
  if (uniqueNames.size !== allRuleNames.length) {
    const duplicates = allRuleNames.filter((name, index) => allRuleNames.indexOf(name) !== index);
    ctx.addIssue({
      code: z.ZodIssueCode.custom,
      message: `Conflicting rule definitions found: ${duplicates.join(", ")}. Rules must be unique! Please check your plugins.`
    });
  }
}
function validateRuleOptions(value, ctx) {
  const ruleToOptions = {};
  value.forEach((configObject) => {
    if (isConfigObject(configObject)) {
      Object.entries(configObject.rules).forEach(([ruleName, severityOrTuple]) => {
        const prevOptions = ruleToOptions[ruleName];
        const ruleOptions = getOptions(severityOrTuple);
        if (!prevOptions) {
          ruleToOptions[ruleName] = ruleOptions;
          return;
        }
        if (ruleOptions && prevOptions && !isEqual(ruleOptions, prevOptions)) {
          ctx.addIssue({
            code: z.ZodIssueCode.custom,
            message: `
                Rule "${ruleName}" has multiple options provided!
                  ${JSON.stringify(ruleToOptions[ruleName])}
                and
                  ${JSON.stringify(ruleOptions)}.
                You can only provide options for a rule once.`
          });
        }
      });
    }
  });
}
function buildValidationScheme(rawConfig) {
  const allRuleNames = getAllRuleNames(rawConfig.filter(isPlugin));
  if (allRuleNames.length === 0) {
    throw new Error(NO_RULES_ERROR_MESSAGE);
  }
  const ruleResultScheme = z.object({
    diagnostics: z.array(z.any())
  });
  return z.array(
    z.union([
      z.object({
        ignores: z.array(z.string())
      }).passthrough(),
      z.object({
        files: z.optional(z.array(z.string())),
        ignores: z.optional(z.array(z.string())),
        // zod.record requires at least one element in the array, so we need "as [string, ...string[]]"
        rules: z.record(
          z.enum(allRuleNames),
          z.union([
            z.enum(["off", "error", "warn"]),
            z.tuple([z.enum(["error", "warn"]), z.object({}).passthrough()])
          ])
        )
      }),
      z.object({
        meta: z.object({
          name: z.string(),
          version: z.string()
        }),
        ruleDefinitions: z.array(
          z.object({
            name: z.string(),
            check: z.function().args().returns(z.union([z.promise(ruleResultScheme), ruleResultScheme]))
          })
        )
      })
    ])
  ).superRefine(validateConfigObjectsNumber).superRefine(validateRuleOptions).superRefine(validateRuleUniqueness);
}
function validateConfig(rawConfig) {
  const isOldConfig = typeof rawConfig === "object" && !Array.isArray(rawConfig);
  const isWrongShape = !Array.isArray(rawConfig);
  if (isOldConfig) {
    throw new Error(OLD_CONFIG_ERROR_MESSAGE);
  }
  if (isWrongShape) {
    throw new Error(WRONG_CONFIG_SHAPE_ERROR_MESSAGE);
  }
  const validationScheme = buildValidationScheme(rawConfig);
  return validationScheme.parse(rawConfig);
}

// src/models/config/transform-globs.ts
init_cjs_shim();

// src/shared/globs/index.ts
init_cjs_shim();

// src/shared/globs/not.ts
init_cjs_shim();
function not2(globs) {
  return {
    files: globs.files,
    ignores: globs.ignores,
    inverted: true
  };
}

// src/shared/globs/utilities.ts
init_cjs_shim();
function isNegatedGlob(pattern) {
  return pattern.startsWith("!");
}
function getGlobPath(pattern) {
  return isNegatedGlob(pattern) ? pattern.slice(1) : pattern;
}
function replaceGlobPath(pattern, replacement) {
  const sanitizedReplacement = getGlobPath(replacement);
  return isNegatedGlob(pattern) ? `!${sanitizedReplacement}` : sanitizedReplacement;
}

// src/shared/globs/apply-exclusion.ts
init_cjs_shim();

// src/shared/globs/create-filter-according-to-globs.ts
init_cjs_shim();
function createFilterAccordingToGlobs({ inclusions, exclusions }) {
  const thereAreInclusions = Array.isArray(inclusions);
  const thereAreExclusions = Array.isArray(exclusions);
  const inclusionsEmpty = thereAreInclusions && inclusions.length === 0;
  function filterAccordingToGlobs(path2) {
    const matchesInclusionPatterns = !thereAreInclusions || inclusions.some((pattern) => micromatch.isMatch(path2, pattern));
    let isIgnored = false;
    if (inclusionsEmpty) {
      return false;
    }
    if (matchesInclusionPatterns && thereAreExclusions) {
      isIgnored = exclusions.filter((pattern) => !isNegatedGlob(pattern)).some((pattern) => micromatch.isMatch(path2, pattern));
      if (isIgnored) {
        const isNegated = exclusions.some(
          (ignorePattern) => isNegatedGlob(ignorePattern) && micromatch.isMatch(path2, ignorePattern.slice(1))
        );
        isIgnored = !isNegated;
      }
    }
    return matchesInclusionPatterns && !isIgnored;
  }
  return filterAccordingToGlobs;
}

// src/shared/globs/apply-exclusion.ts
function removeEmptyFolders(node) {
  const children = node.children.map((node2) => node2.type === "folder" ? removeEmptyFolders(node2) : node2).filter((node2) => node2.type === "folder" ? node2.children.length > 0 : true);
  return {
    ...node,
    children
  };
}
function copyNode(node, deep = false) {
  if (node.type === "folder") {
    const newChildren = deep ? node.children.map((child) => child.type === "folder" ? copyNode(child, true) : child) : [];
    return {
      ...node,
      children: newChildren
    };
  }
  return { ...node };
}
function excludeFilesBasedOnGlobs(vfs, globs) {
  const vfsCopy = copyNode(vfs, true);
  function isIncluded(path2) {
    return globs.reduce((prev, { files, ignores, ...rest }) => {
      const invertedGlob = "inverted" in rest;
      const matches = createFilterAccordingToGlobs({ inclusions: files, exclusions: ignores });
      const current = matches(path2);
      if (invertedGlob && current) {
        return false;
      } else if (invertedGlob && !current || !invertedGlob && !current) {
        return prev;
      } else {
        return current;
      }
    }, false);
  }
  function walk(node) {
    return {
      ...node,
      children: node.children.filter(({ type, path: path2 }) => type === "folder" ? true : isIncluded(path2)).map((child) => {
        if (child.type === "folder") {
          return walk(child);
        }
        return child;
      })
    };
  }
  return walk(vfsCopy);
}
function applyExclusion(vfs, globs) {
  const vfsCopy = copyNode(vfs, true);
  return removeEmptyFolders(excludeFilesBasedOnGlobs(vfsCopy, globs));
}

// src/models/config/transform-globs.ts
function convertRelativeGlobsToAbsolute(rootPath, globs) {
  const needsConversion = (glob) => !isAbsolute(glob);
  return globs.map((originalGlob) => {
    const globClearPath = getGlobPath(originalGlob);
    return needsConversion(globClearPath) ? replaceGlobPath(originalGlob, resolve(rootPath, globClearPath).replace(/\\/g, "/")) : originalGlob;
  });
}
function stripTrailingSlashes(globs) {
  return globs.map((ignore) => ignore.replace(/\/$/, ""));
}
function transformGlobs(config, configLocationPath) {
  if (!configLocationPath) {
    return config;
  }
  const globsTransformationPipeline = function(globs) {
    const convertToAbsolute = (globs2) => convertRelativeGlobsToAbsolute(configLocationPath, globs2);
    return convertToAbsolute(stripTrailingSlashes(globs));
  };
  return config.map((item) => {
    if (!isConfiguration(item)) {
      return item;
    }
    const newItem = {
      ...item
    };
    if (newItem.ignores) {
      newItem.ignores = globsTransformationPipeline(newItem.ignores);
    }
    if (isConfigObject(newItem) && newItem.files) {
      newItem.files = globsTransformationPipeline(newItem.files);
    }
    return newItem;
  });
}

// src/models/config/index.ts
var $ruleInstructions = createStore(null);
var setRuleInstructions = createEvent();
$ruleInstructions.on(setRuleInstructions, (_state, payload) => payload);
var $globalIgnores = createStore([]);
var setGlobalIgnores = createEvent();
$globalIgnores.on(setGlobalIgnores, (_state, payload) => payload);
var $plugins = createStore([]);
var setPlugins = createEvent();
$plugins.on(setPlugins, (_state, payload) => payload);
var $enabledRules = combine($ruleInstructions, $plugins, (ruleInstructions, plugins) => {
  const rulesThatHaveInstructions = ruleInstructions ? Object.keys(ruleInstructions) : [];
  const allRules = plugins.flatMap((plugin) => plugin.ruleDefinitions);
  return allRules.filter((rule) => rulesThatHaveInstructions.includes(rule.name));
});
function processConfiguration(rawConfig, configLocationFolder) {
  const validatedConfig = validateConfig(rawConfig);
  const plugins = rawConfig.filter(isPlugin);
  const configTransformedGlobs = transformGlobs(validatedConfig, configLocationFolder);
  const ruleInstructions = createRuleInstructions(configTransformedGlobs);
  setPlugins(plugins);
  setGlobalIgnores(configTransformedGlobs.filter(isGlobalIgnore));
  setRuleInstructions(ruleInstructions);
  return validatedConfig;
}
function getEnabledRules() {
  return $enabledRules.getState();
}
function getRuleOptions(ruleName) {
  return $ruleInstructions.getState()?.[ruleName].options || null;
}
function getGlobalIgnores() {
  return $globalIgnores.getState();
}
function getGlobsForRule(ruleName) {
  return $ruleInstructions.getState()?.[ruleName].globGroups || [];
}

// src/features/run-rule/index.ts
init_cjs_shim();

// src/features/run-rule/run-rule.ts
init_cjs_shim();

// src/features/run-rule/prepare-vfs-for-rule-run.ts
init_cjs_shim();
function prepareVfsForRuleRun(vfs, globGroups) {
  const globs = globGroups.map(
    ({ files, ignores, severity }) => severity === "off" ? not2({ files, ignores }) : { files, ignores }
  );
  return applyExclusion(vfs, globs);
}

// src/features/run-rule/run-rule.ts
async function runRule(vfs, rule) {
  const globsForRule = getGlobsForRule(rule.name);
  const finalVfs = prepareVfsForRuleRun(vfs, globsForRule);
  if (!finalVfs) {
    return Promise.resolve({ diagnostics: [] });
  }
  return Promise.resolve(rule.check(finalVfs, getRuleOptions(rule.name) ?? {}));
}

// src/features/remove-global-ignores-from-vfs/index.ts
init_cjs_shim();

// src/features/remove-global-ignores-from-vfs/remove-global-ignores-from-vfs.ts
init_cjs_shim();

// src/shared/globs/merge-glob-groups.ts
init_cjs_shim();
function mergeGlobGroups(globs) {
  return globs.reduce((acc, { files, ignores }) => {
    const interimGlobGroup = { ...acc };
    if (files) {
      interimGlobGroup.files = [...acc.files || [], ...files];
    }
    if (ignores) {
      interimGlobGroup.ignores = [...acc.ignores || [], ...ignores];
    }
    return interimGlobGroup;
  });
}

// src/features/remove-global-ignores-from-vfs/remove-global-ignores-from-vfs.ts
function removeGlobalIgnoreFromVfs(vfs, globalIgnores) {
  if (globalIgnores.length === 0) {
    return vfs;
  }
  const globalIgnoresAsGlobGroup = mergeGlobGroups(globalIgnores);
  return applyExclusion(vfs, [globalIgnoresAsGlobGroup]);
}

// src/features/calculate-diagnostic-severities/index.ts
init_cjs_shim();

// src/features/calculate-diagnostic-severities/calculate-final-severity.ts
init_cjs_shim();

// src/shared/file-system.ts
init_cjs_shim();
function isPathInTree(vfs, paths) {
  const pathsToCheck = Array.isArray(paths) ? paths : [paths];
  const results = new Array(pathsToCheck.length).fill(false);
  let found = 0;
  const stack = [vfs];
  while (stack.length > 0) {
    const node = stack.pop();
    const currentPathIndex = pathsToCheck.indexOf(node.path);
    if (currentPathIndex !== -1) {
      results[currentPathIndex] = true;
      found++;
    }
    if (found === pathsToCheck.length) {
      break;
    }
    if (node.type === "folder") {
      for (let i = 0; i < node.children.length; i++) {
        stack.push(node.children[i]);
      }
    }
  }
  return typeof paths === "string" ? results[0] : results;
}
async function existsAndIsFolder(path2) {
  try {
    await access(path2);
    return (await stat(path2)).isDirectory();
  } catch {
    return false;
  }
}

// src/features/calculate-diagnostic-severities/calculate-final-severity.ts
function calculateFinalSeverities(vfs, ruleName, paths) {
  const globGroups = getGlobsForRule(ruleName);
  const errorGlobs = globGroups.map(({ severity, ...rest }) => severity === "error" ? { ...rest } : not2({ ...rest }));
  const errorVfs = applyExclusion(vfs, errorGlobs);
  return isPathInTree(errorVfs, paths).map((isError) => isError ? "error" : "warn");
}

// src/app.ts
function getRuleDescriptionUrl(ruleName) {
  const withoutNamespace = ruleName.split("/")[1];
  return new URL(
    `https://github.com/feature-sliced/steiger/tree/master/packages/steiger-plugin-fsd/src/${withoutNamespace}`
  );
}
async function runRules({ vfs, rules }) {
  const vfsWithoutGlobalIgnores = removeGlobalIgnoreFromVfs(vfs, getGlobalIgnores());
  const ruleResults = await Promise.all(rules.map((rule) => runRule(vfsWithoutGlobalIgnores, rule)));
  return ruleResults.flatMap(({ diagnostics }, ruleResultsIndex) => {
    const ruleName = rules[ruleResultsIndex].name;
    const severities = calculateFinalSeverities(
      vfsWithoutGlobalIgnores,
      ruleName,
      diagnostics.map((d) => d.location.path)
    );
    return diagnostics.sort((a, b) => a.location.path.localeCompare(b.location.path)).map((d, index) => ({
      ...d,
      ruleName,
      getRuleDescriptionUrl,
      severity: severities[index]
    }));
  });
}
var linter = {
  run: (path2) => scan(path2).then(
    (vfs) => runRules({
      vfs,
      rules: getEnabledRules()
    })
  ),
  watch: async (path2) => {
    const { vfs, watcher } = await createWatcher(path2);
    const treeChanged = debounce(vfs.$tree, 500);
    const runRulesFx = createEffect(runRules);
    sample({
      clock: defer({ clock: [treeChanged, $enabledRules], until: not(runRulesFx.pending) }),
      source: {
        vfs: vfs.$tree,
        rules: $enabledRules
      },
      target: runRulesFx
    });
    return [runRulesFx.doneData, () => watcher.close()];
  }
};
function defineConfig(config) {
  return config;
}

export { $plugins, __commonJS, __require, __toESM, defineConfig, existsAndIsFolder, init_cjs_shim, linter, processConfiguration };
