// steiger/dist/cli.mjs
#!/usr/bin/env node
import { __commonJS, init_cjs_shim, __require, __toESM, processConfiguration, $plugins, existsAndIsFolder, linter } from './chunk-BLZHJTX3.mjs';
import { dirname, relative, resolve, parse, join, sep } from 'path';
import * as process3 from 'process';
import process3__default from 'process';
import yargs from 'yargs';
import prexit from 'prexit';
import { hideBin } from 'yargs/helpers';
import pc from 'picocolors';
import { fromError } from 'zod-validation-error';
import { cosmiconfig } from 'cosmiconfig';
import { readdir, open, rm, mkdir, rename } from 'fs/promises';
import { confirm, select, isCancel, outro } from '@clack/prompts';
import { isGitIgnored } from 'globby';
import * as find from 'empathic/find';
import { distance } from 'fastest-levenshtein';
import fsd from '@feature-sliced/steiger-plugin';

// ../../node_modules/.pnpm/has-flag@4.0.0/node_modules/has-flag/index.js
var require_has_flag = __commonJS({
  "../../node_modules/.pnpm/has-flag@4.0.0/node_modules/has-flag/index.js"(exports, module) {
    init_cjs_shim();
    module.exports = (flag, argv2 = process.argv) => {
      const prefix = flag.startsWith("-") ? "" : flag.length === 1 ? "-" : "--";
      const position = argv2.indexOf(prefix + flag);
      const terminatorPosition = argv2.indexOf("--");
      return position !== -1 && (terminatorPosition === -1 || position < terminatorPosition);
    };
  }
});

// ../../node_modules/.pnpm/supports-color@7.2.0/node_modules/supports-color/index.js
var require_supports_color = __commonJS({
  "../../node_modules/.pnpm/supports-color@7.2.0/node_modules/supports-color/index.js"(exports, module) {
    init_cjs_shim();
    var os = __require("os");
    var tty = __require("tty");
    var hasFlag = require_has_flag();
    var { env } = process;
    var forceColor;
    if (hasFlag("no-color") || hasFlag("no-colors") || hasFlag("color=false") || hasFlag("color=never")) {
      forceColor = 0;
    } else if (hasFlag("color") || hasFlag("colors") || hasFlag("color=true") || hasFlag("color=always")) {
      forceColor = 1;
    }
    if ("FORCE_COLOR" in env) {
      if (env.FORCE_COLOR === "true") {
        forceColor = 1;
      } else if (env.FORCE_COLOR === "false") {
        forceColor = 0;
      } else {
        forceColor = env.FORCE_COLOR.length === 0 ? 1 : Math.min(parseInt(env.FORCE_COLOR, 10), 3);
      }
    }
    function translateLevel(level) {
      if (level === 0) {
        return false;
      }
      return {
        level,
        hasBasic: true,
        has256: level >= 2,
        has16m: level >= 3
      };
    }
    function supportsColor(haveStream, streamIsTTY) {
      if (forceColor === 0) {
        return 0;
      }
      if (hasFlag("color=16m") || hasFlag("color=full") || hasFlag("color=truecolor")) {
        return 3;
      }
      if (hasFlag("color=256")) {
        return 2;
      }
      if (haveStream && !streamIsTTY && forceColor === void 0) {
        return 0;
      }
      const min = forceColor || 0;
      if (env.TERM === "dumb") {
        return min;
      }
      if (process.platform === "win32") {
        const osRelease = os.release().split(".");
        if (Number(osRelease[0]) >= 10 && Number(osRelease[2]) >= 10586) {
          return Number(osRelease[2]) >= 14931 ? 3 : 2;
        }
        return 1;
      }
      if ("CI" in env) {
        if (["TRAVIS", "CIRCLECI", "APPVEYOR", "GITLAB_CI", "GITHUB_ACTIONS", "BUILDKITE"].some((sign) => sign in env) || env.CI_NAME === "codeship") {
          return 1;
        }
        return min;
      }
      if ("TEAMCITY_VERSION" in env) {
        return /^(9\.(0*[1-9]\d*)\.|\d{2,}\.)/.test(env.TEAMCITY_VERSION) ? 1 : 0;
      }
      if (env.COLORTERM === "truecolor") {
        return 3;
      }
      if ("TERM_PROGRAM" in env) {
        const version = parseInt((env.TERM_PROGRAM_VERSION || "").split(".")[0], 10);
        switch (env.TERM_PROGRAM) {
          case "iTerm.app":
            return version >= 3 ? 3 : 2;
          case "Apple_Terminal":
            return 2;
        }
      }
      if (/-256(color)?$/i.test(env.TERM)) {
        return 2;
      }
      if (/^screen|^xterm|^vt100|^vt220|^rxvt|color|ansi|cygwin|linux/i.test(env.TERM)) {
        return 1;
      }
      if ("COLORTERM" in env) {
        return 1;
      }
      return min;
    }
    function getSupportLevel(stream) {
      const level = supportsColor(stream, stream && stream.isTTY);
      return translateLevel(level);
    }
    module.exports = {
      supportsColor: getSupportLevel,
      stdout: translateLevel(supportsColor(true, tty.isatty(1))),
      stderr: translateLevel(supportsColor(true, tty.isatty(2)))
    };
  }
});

// ../../node_modules/.pnpm/supports-hyperlinks@2.3.0/node_modules/supports-hyperlinks/index.js
var require_supports_hyperlinks = __commonJS({
  "../../node_modules/.pnpm/supports-hyperlinks@2.3.0/node_modules/supports-hyperlinks/index.js"(exports, module) {
    init_cjs_shim();
    var supportsColor = require_supports_color();
    var hasFlag = require_has_flag();
    function parseVersion(versionString) {
      if (/^\d{3,4}$/.test(versionString)) {
        const m = /(\d{1,2})(\d{2})/.exec(versionString);
        return {
          major: 0,
          minor: parseInt(m[1], 10),
          patch: parseInt(m[2], 10)
        };
      }
      const versions = (versionString || "").split(".").map((n) => parseInt(n, 10));
      return {
        major: versions[0],
        minor: versions[1],
        patch: versions[2]
      };
    }
    function supportsHyperlink(stream) {
      const { env } = process;
      if ("FORCE_HYPERLINK" in env) {
        return !(env.FORCE_HYPERLINK.length > 0 && parseInt(env.FORCE_HYPERLINK, 10) === 0);
      }
      if (hasFlag("no-hyperlink") || hasFlag("no-hyperlinks") || hasFlag("hyperlink=false") || hasFlag("hyperlink=never")) {
        return false;
      }
      if (hasFlag("hyperlink=true") || hasFlag("hyperlink=always")) {
        return true;
      }
      if ("NETLIFY" in env) {
        return true;
      }
      if (!supportsColor.supportsColor(stream)) {
        return false;
      }
      if (stream && !stream.isTTY) {
        return false;
      }
      if (process.platform === "win32") {
        return false;
      }
      if ("CI" in env) {
        return false;
      }
      if ("TEAMCITY_VERSION" in env) {
        return false;
      }
      if ("TERM_PROGRAM" in env) {
        const version = parseVersion(env.TERM_PROGRAM_VERSION);
        switch (env.TERM_PROGRAM) {
          case "iTerm.app":
            if (version.major === 3) {
              return version.minor >= 1;
            }
            return version.major > 3;
          case "WezTerm":
            return version.major >= 20200620;
          case "vscode":
            return version.major > 1 || version.major === 1 && version.minor >= 72;
        }
      }
      if ("VTE_VERSION" in env) {
        if (env.VTE_VERSION === "0.50.0") {
          return false;
        }
        const version = parseVersion(env.VTE_VERSION);
        return version.major > 0 || version.minor >= 50;
      }
      return false;
    }
    module.exports = {
      supportsHyperlink,
      stdout: supportsHyperlink(process.stdout),
      stderr: supportsHyperlink(process.stderr)
    };
  }
});

// src/cli.ts
init_cjs_shim();

// ../pretty-reporter/src/index.ts
init_cjs_shim();

// ../../node_modules/.pnpm/figures@6.1.0/node_modules/figures/index.js
init_cjs_shim();

// ../../node_modules/.pnpm/is-unicode-supported@2.1.0/node_modules/is-unicode-supported/index.js
init_cjs_shim();
function isUnicodeSupported() {
  const { env } = process3__default;
  const { TERM, TERM_PROGRAM } = env;
  if (process3__default.platform !== "win32") {
    return TERM !== "linux";
  }
  return Boolean(env.WT_SESSION) || Boolean(env.TERMINUS_SUBLIME) || env.ConEmuTask === "{cmd::Cmder}" || TERM_PROGRAM === "Terminus-Sublime" || TERM_PROGRAM === "vscode" || TERM === "xterm-256color" || TERM === "alacritty" || TERM === "rxvt-unicode" || TERM === "rxvt-unicode-256color" || env.TERMINAL_EMULATOR === "JetBrains-JediTerm";
}

// ../../node_modules/.pnpm/figures@6.1.0/node_modules/figures/index.js
var common = {
  circleQuestionMark: "(?)",
  questionMarkPrefix: "(?)",
  square: "\u2588",
  squareDarkShade: "\u2593",
  squareMediumShade: "\u2592",
  squareLightShade: "\u2591",
  squareTop: "\u2580",
  squareBottom: "\u2584",
  squareLeft: "\u258C",
  squareRight: "\u2590",
  squareCenter: "\u25A0",
  bullet: "\u25CF",
  dot: "\u2024",
  ellipsis: "\u2026",
  pointerSmall: "\u203A",
  triangleUp: "\u25B2",
  triangleUpSmall: "\u25B4",
  triangleDown: "\u25BC",
  triangleDownSmall: "\u25BE",
  triangleLeftSmall: "\u25C2",
  triangleRightSmall: "\u25B8",
  home: "\u2302",
  heart: "\u2665",
  musicNote: "\u266A",
  musicNoteBeamed: "\u266B",
  arrowUp: "\u2191",
  arrowDown: "\u2193",
  arrowLeft: "\u2190",
  arrowRight: "\u2192",
  arrowLeftRight: "\u2194",
  arrowUpDown: "\u2195",
  almostEqual: "\u2248",
  notEqual: "\u2260",
  lessOrEqual: "\u2264",
  greaterOrEqual: "\u2265",
  identical: "\u2261",
  infinity: "\u221E",
  subscriptZero: "\u2080",
  subscriptOne: "\u2081",
  subscriptTwo: "\u2082",
  subscriptThree: "\u2083",
  subscriptFour: "\u2084",
  subscriptFive: "\u2085",
  subscriptSix: "\u2086",
  subscriptSeven: "\u2087",
  subscriptEight: "\u2088",
  subscriptNine: "\u2089",
  oneHalf: "\xBD",
  oneThird: "\u2153",
  oneQuarter: "\xBC",
  oneFifth: "\u2155",
  oneSixth: "\u2159",
  oneEighth: "\u215B",
  twoThirds: "\u2154",
  twoFifths: "\u2156",
  threeQuarters: "\xBE",
  threeFifths: "\u2157",
  threeEighths: "\u215C",
  fourFifths: "\u2158",
  fiveSixths: "\u215A",
  fiveEighths: "\u215D",
  sevenEighths: "\u215E",
  line: "\u2500",
  lineBold: "\u2501",
  lineDouble: "\u2550",
  lineDashed0: "\u2504",
  lineDashed1: "\u2505",
  lineDashed2: "\u2508",
  lineDashed3: "\u2509",
  lineDashed4: "\u254C",
  lineDashed5: "\u254D",
  lineDashed6: "\u2574",
  lineDashed7: "\u2576",
  lineDashed8: "\u2578",
  lineDashed9: "\u257A",
  lineDashed10: "\u257C",
  lineDashed11: "\u257E",
  lineDashed12: "\u2212",
  lineDashed13: "\u2013",
  lineDashed14: "\u2010",
  lineDashed15: "\u2043",
  lineVertical: "\u2502",
  lineVerticalBold: "\u2503",
  lineVerticalDouble: "\u2551",
  lineVerticalDashed0: "\u2506",
  lineVerticalDashed1: "\u2507",
  lineVerticalDashed2: "\u250A",
  lineVerticalDashed3: "\u250B",
  lineVerticalDashed4: "\u254E",
  lineVerticalDashed5: "\u254F",
  lineVerticalDashed6: "\u2575",
  lineVerticalDashed7: "\u2577",
  lineVerticalDashed8: "\u2579",
  lineVerticalDashed9: "\u257B",
  lineVerticalDashed10: "\u257D",
  lineVerticalDashed11: "\u257F",
  lineDownLeft: "\u2510",
  lineDownLeftArc: "\u256E",
  lineDownBoldLeftBold: "\u2513",
  lineDownBoldLeft: "\u2512",
  lineDownLeftBold: "\u2511",
  lineDownDoubleLeftDouble: "\u2557",
  lineDownDoubleLeft: "\u2556",
  lineDownLeftDouble: "\u2555",
  lineDownRight: "\u250C",
  lineDownRightArc: "\u256D",
  lineDownBoldRightBold: "\u250F",
  lineDownBoldRight: "\u250E",
  lineDownRightBold: "\u250D",
  lineDownDoubleRightDouble: "\u2554",
  lineDownDoubleRight: "\u2553",
  lineDownRightDouble: "\u2552",
  lineUpLeft: "\u2518",
  lineUpLeftArc: "\u256F",
  lineUpBoldLeftBold: "\u251B",
  lineUpBoldLeft: "\u251A",
  lineUpLeftBold: "\u2519",
  lineUpDoubleLeftDouble: "\u255D",
  lineUpDoubleLeft: "\u255C",
  lineUpLeftDouble: "\u255B",
  lineUpRight: "\u2514",
  lineUpRightArc: "\u2570",
  lineUpBoldRightBold: "\u2517",
  lineUpBoldRight: "\u2516",
  lineUpRightBold: "\u2515",
  lineUpDoubleRightDouble: "\u255A",
  lineUpDoubleRight: "\u2559",
  lineUpRightDouble: "\u2558",
  lineUpDownLeft: "\u2524",
  lineUpBoldDownBoldLeftBold: "\u252B",
  lineUpBoldDownBoldLeft: "\u2528",
  lineUpDownLeftBold: "\u2525",
  lineUpBoldDownLeftBold: "\u2529",
  lineUpDownBoldLeftBold: "\u252A",
  lineUpDownBoldLeft: "\u2527",
  lineUpBoldDownLeft: "\u2526",
  lineUpDoubleDownDoubleLeftDouble: "\u2563",
  lineUpDoubleDownDoubleLeft: "\u2562",
  lineUpDownLeftDouble: "\u2561",
  lineUpDownRight: "\u251C",
  lineUpBoldDownBoldRightBold: "\u2523",
  lineUpBoldDownBoldRight: "\u2520",
  lineUpDownRightBold: "\u251D",
  lineUpBoldDownRightBold: "\u2521",
  lineUpDownBoldRightBold: "\u2522",
  lineUpDownBoldRight: "\u251F",
  lineUpBoldDownRight: "\u251E",
  lineUpDoubleDownDoubleRightDouble: "\u2560",
  lineUpDoubleDownDoubleRight: "\u255F",
  lineUpDownRightDouble: "\u255E",
  lineDownLeftRight: "\u252C",
  lineDownBoldLeftBoldRightBold: "\u2533",
  lineDownLeftBoldRightBold: "\u252F",
  lineDownBoldLeftRight: "\u2530",
  lineDownBoldLeftBoldRight: "\u2531",
  lineDownBoldLeftRightBold: "\u2532",
  lineDownLeftRightBold: "\u252E",
  lineDownLeftBoldRight: "\u252D",
  lineDownDoubleLeftDoubleRightDouble: "\u2566",
  lineDownDoubleLeftRight: "\u2565",
  lineDownLeftDoubleRightDouble: "\u2564",
  lineUpLeftRight: "\u2534",
  lineUpBoldLeftBoldRightBold: "\u253B",
  lineUpLeftBoldRightBold: "\u2537",
  lineUpBoldLeftRight: "\u2538",
  lineUpBoldLeftBoldRight: "\u2539",
  lineUpBoldLeftRightBold: "\u253A",
  lineUpLeftRightBold: "\u2536",
  lineUpLeftBoldRight: "\u2535",
  lineUpDoubleLeftDoubleRightDouble: "\u2569",
  lineUpDoubleLeftRight: "\u2568",
  lineUpLeftDoubleRightDouble: "\u2567",
  lineUpDownLeftRight: "\u253C",
  lineUpBoldDownBoldLeftBoldRightBold: "\u254B",
  lineUpDownBoldLeftBoldRightBold: "\u2548",
  lineUpBoldDownLeftBoldRightBold: "\u2547",
  lineUpBoldDownBoldLeftRightBold: "\u254A",
  lineUpBoldDownBoldLeftBoldRight: "\u2549",
  lineUpBoldDownLeftRight: "\u2540",
  lineUpDownBoldLeftRight: "\u2541",
  lineUpDownLeftBoldRight: "\u253D",
  lineUpDownLeftRightBold: "\u253E",
  lineUpBoldDownBoldLeftRight: "\u2542",
  lineUpDownLeftBoldRightBold: "\u253F",
  lineUpBoldDownLeftBoldRight: "\u2543",
  lineUpBoldDownLeftRightBold: "\u2544",
  lineUpDownBoldLeftBoldRight: "\u2545",
  lineUpDownBoldLeftRightBold: "\u2546",
  lineUpDoubleDownDoubleLeftDoubleRightDouble: "\u256C",
  lineUpDoubleDownDoubleLeftRight: "\u256B",
  lineUpDownLeftDoubleRightDouble: "\u256A",
  lineCross: "\u2573",
  lineBackslash: "\u2572",
  lineSlash: "\u2571"
};
var specialMainSymbols = {
  tick: "\u2714",
  info: "\u2139",
  warning: "\u26A0",
  cross: "\u2718",
  squareSmall: "\u25FB",
  squareSmallFilled: "\u25FC",
  circle: "\u25EF",
  circleFilled: "\u25C9",
  circleDotted: "\u25CC",
  circleDouble: "\u25CE",
  circleCircle: "\u24DE",
  circleCross: "\u24E7",
  circlePipe: "\u24BE",
  radioOn: "\u25C9",
  radioOff: "\u25EF",
  checkboxOn: "\u2612",
  checkboxOff: "\u2610",
  checkboxCircleOn: "\u24E7",
  checkboxCircleOff: "\u24BE",
  pointer: "\u276F",
  triangleUpOutline: "\u25B3",
  triangleLeft: "\u25C0",
  triangleRight: "\u25B6",
  lozenge: "\u25C6",
  lozengeOutline: "\u25C7",
  hamburger: "\u2630",
  smiley: "\u32E1",
  mustache: "\u0DF4",
  star: "\u2605",
  play: "\u25B6",
  nodejs: "\u2B22",
  oneSeventh: "\u2150",
  oneNinth: "\u2151",
  oneTenth: "\u2152"
};
var specialFallbackSymbols = {
  tick: "\u221A",
  info: "i",
  warning: "\u203C",
  cross: "\xD7",
  squareSmall: "\u25A1",
  squareSmallFilled: "\u25A0",
  circle: "( )",
  circleFilled: "(*)",
  circleDotted: "( )",
  circleDouble: "( )",
  circleCircle: "(\u25CB)",
  circleCross: "(\xD7)",
  circlePipe: "(\u2502)",
  radioOn: "(*)",
  radioOff: "( )",
  checkboxOn: "[\xD7]",
  checkboxOff: "[ ]",
  checkboxCircleOn: "(\xD7)",
  checkboxCircleOff: "( )",
  pointer: ">",
  triangleUpOutline: "\u2206",
  triangleLeft: "\u25C4",
  triangleRight: "\u25BA",
  lozenge: "\u2666",
  lozengeOutline: "\u25CA",
  hamburger: "\u2261",
  smiley: "\u263A",
  mustache: "\u250C\u2500\u2510",
  star: "\u2736",
  play: "\u25BA",
  nodejs: "\u2666",
  oneSeventh: "1/7",
  oneNinth: "1/9",
  oneTenth: "1/10"
};
var mainSymbols = { ...common, ...specialMainSymbols };
var fallbackSymbols = { ...common, ...specialFallbackSymbols };
var shouldUseMain = isUnicodeSupported();
var figures = shouldUseMain ? mainSymbols : fallbackSymbols;
var figures_default = figures;

// ../pretty-reporter/src/format-single-diagnostic.ts
init_cjs_shim();

// ../../node_modules/.pnpm/terminal-link@3.0.0/node_modules/terminal-link/index.js
init_cjs_shim();

// ../../node_modules/.pnpm/ansi-escapes@5.0.0/node_modules/ansi-escapes/index.js
init_cjs_shim();
var ESC = "\x1B[";
var OSC = "\x1B]";
var BEL = "\x07";
var SEP = ";";
var isTerminalApp = process.env.TERM_PROGRAM === "Apple_Terminal";
var ansiEscapes = {};
ansiEscapes.cursorTo = (x, y) => {
  if (typeof x !== "number") {
    throw new TypeError("The `x` argument is required");
  }
  if (typeof y !== "number") {
    return ESC + (x + 1) + "G";
  }
  return ESC + (y + 1) + ";" + (x + 1) + "H";
};
ansiEscapes.cursorMove = (x, y) => {
  if (typeof x !== "number") {
    throw new TypeError("The `x` argument is required");
  }
  let returnValue = "";
  if (x < 0) {
    returnValue += ESC + -x + "D";
  } else if (x > 0) {
    returnValue += ESC + x + "C";
  }
  if (y < 0) {
    returnValue += ESC + -y + "A";
  } else if (y > 0) {
    returnValue += ESC + y + "B";
  }
  return returnValue;
};
ansiEscapes.cursorUp = (count = 1) => ESC + count + "A";
ansiEscapes.cursorDown = (count = 1) => ESC + count + "B";
ansiEscapes.cursorForward = (count = 1) => ESC + count + "C";
ansiEscapes.cursorBackward = (count = 1) => ESC + count + "D";
ansiEscapes.cursorLeft = ESC + "G";
ansiEscapes.cursorSavePosition = isTerminalApp ? "\x1B7" : ESC + "s";
ansiEscapes.cursorRestorePosition = isTerminalApp ? "\x1B8" : ESC + "u";
ansiEscapes.cursorGetPosition = ESC + "6n";
ansiEscapes.cursorNextLine = ESC + "E";
ansiEscapes.cursorPrevLine = ESC + "F";
ansiEscapes.cursorHide = ESC + "?25l";
ansiEscapes.cursorShow = ESC + "?25h";
ansiEscapes.eraseLines = (count) => {
  let clear = "";
  for (let i = 0; i < count; i++) {
    clear += ansiEscapes.eraseLine + (i < count - 1 ? ansiEscapes.cursorUp() : "");
  }
  if (count) {
    clear += ansiEscapes.cursorLeft;
  }
  return clear;
};
ansiEscapes.eraseEndLine = ESC + "K";
ansiEscapes.eraseStartLine = ESC + "1K";
ansiEscapes.eraseLine = ESC + "2K";
ansiEscapes.eraseDown = ESC + "J";
ansiEscapes.eraseUp = ESC + "1J";
ansiEscapes.eraseScreen = ESC + "2J";
ansiEscapes.scrollUp = ESC + "S";
ansiEscapes.scrollDown = ESC + "T";
ansiEscapes.clearScreen = "\x1Bc";
ansiEscapes.clearTerminal = process.platform === "win32" ? `${ansiEscapes.eraseScreen}${ESC}0f` : (
  // 1. Erases the screen (Only done in case `2` is not supported)
  // 2. Erases the whole screen including scrollback buffer
  // 3. Moves cursor to the top-left position
  // More info: https://www.real-world-systems.com/docs/ANSIcode.html
  `${ansiEscapes.eraseScreen}${ESC}3J${ESC}H`
);
ansiEscapes.beep = BEL;
ansiEscapes.link = (text, url) => {
  return [
    OSC,
    "8",
    SEP,
    SEP,
    url,
    BEL,
    text,
    OSC,
    "8",
    SEP,
    SEP,
    BEL
  ].join("");
};
ansiEscapes.image = (buffer, options = {}) => {
  let returnValue = `${OSC}1337;File=inline=1`;
  if (options.width) {
    returnValue += `;width=${options.width}`;
  }
  if (options.height) {
    returnValue += `;height=${options.height}`;
  }
  if (options.preserveAspectRatio === false) {
    returnValue += ";preserveAspectRatio=0";
  }
  return returnValue + ":" + buffer.toString("base64") + BEL;
};
ansiEscapes.iTerm = {
  setCwd: (cwd2 = process.cwd()) => `${OSC}50;CurrentDir=${cwd2}${BEL}`,
  annotation: (message, options = {}) => {
    let returnValue = `${OSC}1337;`;
    const hasX = typeof options.x !== "undefined";
    const hasY = typeof options.y !== "undefined";
    if ((hasX || hasY) && !(hasX && hasY && typeof options.length !== "undefined")) {
      throw new Error("`x`, `y` and `length` must be defined when `x` or `y` is defined");
    }
    message = message.replace(/\|/g, "");
    returnValue += options.isHidden ? "AddHiddenAnnotation=" : "AddAnnotation=";
    if (options.length > 0) {
      returnValue += (hasX ? [message, options.length, options.x, options.y] : [options.length, message]).join("|");
    } else {
      returnValue += message;
    }
    return returnValue + BEL;
  }
};
var ansi_escapes_default = ansiEscapes;

// ../../node_modules/.pnpm/terminal-link@3.0.0/node_modules/terminal-link/index.js
var import_supports_hyperlinks = __toESM(require_supports_hyperlinks(), 1);
function terminalLink(text, url, { target = "stdout", ...options } = {}) {
  if (!import_supports_hyperlinks.default[target]) {
    if (options.fallback === false) {
      return text;
    }
    return typeof options.fallback === "function" ? options.fallback(text, url) : `${text} (\u200B${url}\u200B)`;
  }
  return ansi_escapes_default.link(text, url);
}
terminalLink.isSupported = import_supports_hyperlinks.default.stdout;
terminalLink.stderr = (text, url, options = {}) => terminalLink(text, url, { target: "stderr", ...options });
terminalLink.stderr.isSupported = import_supports_hyperlinks.default.stderr;
function formatSingleDiagnostic(d, cwd2) {
  const x = d.severity === "error" ? pc.red(figures_default.cross) : pc.yellow(figures_default.warning);
  const s2 = pc.reset(figures_default.lineDownRight);
  const bar = pc.reset(figures_default.lineVertical);
  const e = pc.reset(figures_default.lineUpRight);
  const message = pc.reset(d.message);
  const autofixable = d.fixes !== void 0 && d.fixes.length > 0 ? pc.green(`${figures_default.tick} Auto-fixable`) : null;
  const location = pc.underline(formatLocation(d.location, cwd2));
  const ruleName = pc.blue(
    terminalLink(d.ruleName, d.getRuleDescriptionUrl(d.ruleName).toString(), {
      fallback: (text, url) => `${pc.reset(text)}: ${pc.blue(url)}`
    })
  );
  return `
${s2} ${location}
${x} ${message}
${autofixable ? autofixable + `
${bar}` : bar}
${e} ${ruleName}
`.trim();
}
function formatLocation(location, cwd2) {
  let path = relative(cwd2, location.path);
  if (location.line !== void 0) {
    path += `:${location.line}`;
    if (location.column !== void 0) {
      path += `:${location.column}`;
    }
  }
  return path;
}

// ../pretty-reporter/src/pluralization.ts
init_cjs_shim();
function s(amount) {
  return amount === 1 ? "" : "s";
}

// ../pretty-reporter/src/index.ts
function formatPretty(diagnostics, cwd2) {
  if (diagnostics.length === 0) {
    return pc.green(`${figures_default.tick} No problems found!`);
  }
  const errors = diagnostics.filter((d) => d.severity === "error");
  const warnings = diagnostics.filter((d) => d.severity === "warn");
  let footer = "Found " + [
    errors.length > 0 && pc.bold(pc.red(`${errors.length} error${s(errors.length)}`)),
    warnings.length > 0 && pc.bold(pc.yellow(`${warnings.length} warning${s(warnings.length)}`))
  ].filter(Boolean).join(" and ");
  const autofixable = diagnostics.filter((d) => (d.fixes?.length ?? 0) > 0);
  if (autofixable.length === diagnostics.length) {
    footer += ` (all can be fixed automatically with ${pc.bold(pc.green("--fix"))})`;
  } else if (autofixable.length > 0) {
    footer += ` (${autofixable.length} can be fixed automatically with ${pc.bold(pc.green("--fix"))})`;
  } else {
    footer += " (none can be fixed automatically)";
  }
  return "\n" + diagnostics.map((d) => formatSingleDiagnostic(d, cwd2)).join("\n\n") + "\n\n" + // Due to formatting characters, it won't be exactly the size of the footer, that is okay
  pc.gray(figures_default.line.repeat(footer.length)) + "\n " + footer + "\n";
}
function reportPretty(diagnostics, cwd2) {
  console.error(formatPretty(diagnostics, cwd2));
}

// src/features/autofix.ts
init_cjs_shim();
async function applyAutofixes(diagnostics) {
  const stillRelevantDiagnostics = [];
  const fixableDiagnostics = [];
  for (const diagnostic of diagnostics) {
    const fixes = diagnostic.fixes;
    if (!fixes) {
      stillRelevantDiagnostics.push(diagnostic);
      continue;
    }
    fixableDiagnostics.push(diagnostic);
  }
  try {
    await Promise.all(fixableDiagnostics.map(tryToApplyFixes));
  } catch (error) {
    stillRelevantDiagnostics.push(...fixableDiagnostics);
    console.error(error);
  }
  return stillRelevantDiagnostics;
}
async function tryToApplyFixes(diagnostic) {
  const fixes = diagnostic.fixes ?? [];
  return Promise.all(
    fixes.map((fix) => {
      switch (fix.type) {
        case "rename":
          return rename(fix.path, join(dirname(fix.path), fix.newName));
        case "create-file":
          return open(fix.path, "w").then((file) => file.close());
        case "create-folder":
          return mkdir(fix.path, { recursive: true });
        case "delete":
          return rm(fix.path, { recursive: true });
        case "modify-file":
          return open(fix.path, "w").then(async (file) => {
            await file.write(fix.content);
            return file.close();
          });
      }
    })
  );
}

// src/features/choose-root-folder/index.ts
init_cjs_shim();

// src/features/choose-root-folder/choose-from-guesses.ts
init_cjs_shim();

// src/features/choose-root-folder/exit-exception.ts
init_cjs_shim();
var ExitException = class extends Error {
};

// src/features/choose-root-folder/format-command.ts
init_cjs_shim();
function formatCommand(command) {
  return pc.green(`\`${command}\``);
}

// src/features/choose-root-folder/choose-from-guesses.ts
var commonRootFolders = ["src", "app"].map((folder) => `.${sep}${folder}`);
async function chooseFromGuesses() {
  let targetPath2;
  const candidates = await findRootFolderCandidates();
  if (candidates.length === 0) {
    const answer = await confirm({
      message: `You haven't specified a path to check. Would you like to check this folder?`,
      inactive: "No, exit"
    });
    if (answer === true) {
      targetPath2 = ".";
    }
  } else {
    const answer = await select({
      message: `You haven't specified a path to check. Would you like to use one of the following?`,
      options: candidates.map((candidate) => ({ value: candidate, label: candidate })).concat({ value: ".", label: "This folder" }).concat({ value: "", label: "No, exit" })
    });
    if (!isCancel(answer) && answer !== "") {
      targetPath2 = answer;
    }
  }
  if (targetPath2 === void 0) {
    outro(`Alright! To run checks on a specific folder, run ${formatCommand(`steiger .${sep}your-folder`)}.`);
    throw new ExitException();
  } else {
    outro(`Running ${formatCommand(`steiger ${targetPath2}`)}`);
  }
  return targetPath2;
}
async function findRootFolderCandidates() {
  return (await Promise.all(commonRootFolders.map(async (folder) => await existsAndIsFolder(folder) ? folder : void 0))).filter(Boolean);
}

// src/features/choose-root-folder/choose-from-similar.ts
init_cjs_shim();
var typoThreshold = 5;
var gitFolder = find.up(".git");
var isIgnored = await isGitIgnored({ cwd: gitFolder ? dirname(gitFolder) : void 0 });
async function chooseFromSimilar(input) {
  const resolved = relative(".", input);
  const { dir, base } = parse(resolved);
  const existingDir = await resolveWithCorrections(dir || ".");
  const candidates = (await readdir(existingDir, { withFileTypes: true })).filter((entry) => entry.isDirectory() && entry.name !== ".git" && !isIgnored(join(existingDir, entry.name))).map((entry) => entry.name);
  const withDistances = candidates.map((candidate) => [candidate, distance(candidate, base)]);
  const suggestions = withDistances.filter(([_candidate, distance2]) => distance2 <= typoThreshold).sort((a, b) => a[1] - b[1]);
  let answer;
  if (suggestions.length === 1) {
    const confirmation = await confirm({
      message: `${pc.red(input)} is not a folder. Did you mean ${pc.green(`.${sep}${join(existingDir, suggestions[0][0])}`)}?`,
      inactive: "No, exit"
    });
    if (confirmation === true) {
      answer = join(existingDir, suggestions[0][0]);
    } else {
      answer = "";
    }
  } else {
    const selection = await select({
      message: `${pc.red(input)} is not a folder. Did you mean one of the following?`,
      options: suggestions.map(([candidate, _distance]) => ({ value: candidate, label: `.${sep}${join(existingDir, candidate)}` })).concat({ value: "", label: "No, exit" })
    });
    if (selection !== "" && !isCancel(selection)) {
      answer = join(existingDir, selection);
    } else {
      answer = "";
    }
  }
  if (answer !== "") {
    outro(`Running ${formatCommand(`steiger .${sep}${answer}`)}`);
    return answer;
  } else {
    outro(`Alright! To run checks on a specific folder, run ${formatCommand(`steiger .${sep}your-folder`)}.`);
    throw new ExitException();
  }
}
async function resolveWithCorrections(path) {
  let finalPath = ".";
  for (const part of path.split(sep)) {
    if (part === ".") {
      continue;
    } else if (part === "..") {
      finalPath = join(finalPath, part);
    } else {
      const candidates = (await readdir(finalPath, { withFileTypes: true })).filter((entry) => entry.isDirectory()).map((entry) => entry.name);
      const distances = candidates.map((candidate) => distance(candidate, part));
      const bestMatch = candidates[distances.indexOf(Math.min(...distances))];
      finalPath = join(finalPath, bestMatch);
    }
  }
  return finalPath;
}

// package.json
var package_default = {
  version: "0.5.11"};

// src/cli.ts
var { config, filepath } = await cosmiconfig("steiger").search() ?? { config: null, filepath: void 0 };
var defaultConfig = fsd.configs.recommended;
try {
  const configLocationDirectory = filepath ? dirname(filepath) : null;
  processConfiguration(config || defaultConfig, configLocationDirectory);
} catch (err) {
  if (filepath !== void 0) {
    console.error(
      fromError(err, { prefix: `Invalid configuration in ${relative(process3.cwd(), filepath)}` }).toString()
    );
    process3.exit(100);
  }
}
var yargsProgram = yargs(hideBin(process3.argv)).scriptName("steiger").usage("$0 [options] <path>").option("watch", {
  alias: "w",
  demandOption: false,
  describe: "watch filesystem changes",
  type: "boolean"
}).option("fix", {
  demandOption: false,
  describe: "apply auto-fixes",
  type: "boolean"
}).option("fail-on-warnings", {
  demandOption: false,
  describe: "exit with an error code if there are warnings",
  type: "boolean"
}).option("reporter", {
  demandOption: false,
  describe: "specify output format (pretty or json)",
  type: "string",
  choices: ["pretty", "json"],
  default: "pretty"
}).string("_").check((argv2) => {
  const filePaths = argv2._;
  if (filePaths.length > 1) {
    throw new Error("Pass only one path to watch");
  } else {
    return true;
  }
}).help("help", "display help message").alias("help", "h").version(
  [
    package_default.version,
    $plugins.getState().map((plugin) => `${plugin.meta.name}	${plugin.meta.version}`).join("\n")
  ].filter(Boolean).join("\n\n")
).alias("version", "v").showHelpOnFail(true);
var consoleArgs = yargsProgram.parseSync();
var inputPaths = consoleArgs._;
var targetPath;
if (inputPaths.length > 0) {
  if (await existsAndIsFolder(inputPaths[0])) {
    targetPath = resolve(inputPaths[0]);
  } else {
    try {
      targetPath = resolve(await chooseFromSimilar(inputPaths[0]));
    } catch (e) {
      if (e instanceof ExitException) {
        process3.exit(0);
      } else {
        throw e;
      }
    }
  }
} else {
  try {
    targetPath = resolve(await chooseFromGuesses());
  } catch (e) {
    if (e instanceof ExitException) {
      process3.exit(0);
    } else {
      throw e;
    }
  }
}
var printDiagnostics = (diagnostics) => {
  if (consoleArgs.reporter === "json") {
    console.log(JSON.stringify(diagnostics, null, 2));
  } else {
    reportPretty(diagnostics, process3.cwd());
  }
};
if (consoleArgs.watch) {
  const [diagnosticsChanged, stopWatching] = await linter.watch(targetPath);
  const unsubscribe = diagnosticsChanged.watch((state) => {
    console.clear();
    printDiagnostics(state);
    if (consoleArgs.fix) {
      applyAutofixes(state);
    }
  });
  prexit(() => {
    stopWatching();
    unsubscribe();
  });
} else {
  const diagnostics = await linter.run(targetPath);
  let stillRelevantDiagnostics = diagnostics;
  printDiagnostics(diagnostics);
  if (consoleArgs.fix) {
    stillRelevantDiagnostics = await applyAutofixes(diagnostics);
  }
  if (stillRelevantDiagnostics.length > 0) {
    const onlyWarnings = stillRelevantDiagnostics.every((d) => d.severity === "warn");
    if (consoleArgs["fail-on-warnings"] || !onlyWarnings) {
      process3.exit(1);
    }
  }
}
