// @feature-sliced/steiger-plugin/dist/index.js
import { sep, basename, join, relative, resolve, dirname } from 'node:path';
import { getLayers, getSegments, getAllSlices, isSliced, getSlices, isCrossImportPublicApi, layerSequence, unslicedLayers, getIndexes, crossReferenceToken, getAllSegments, isSlice, conventionalSegmentNames, resolveImport } from '@feature-sliced/filesystem';
import * as fs from 'node:fs';
import precinct from 'precinct';
import { parse } from 'tsconfck';
import { partition } from 'lodash-es';
import pluralize from 'pluralize';
import { distance } from 'fastest-levenshtein';

// ../toolkit/dist/index.js
function findAllRecursively(folder, predicate) {
  const result = [];
  function walk(entry) {
    if (predicate(entry)) {
      result.push(entry);
    }
    if (entry.type === "folder") {
      for (const child of entry.children) {
        walk(child);
      }
    }
  }
  walk(folder);
  return result;
}
function enableSpecificRules(plugin2, rulesToEnable, options) {
  const namesOfRulesToEnable = rulesToEnable.map((ruleToEnable) => ruleToEnable.name);
  return [
    plugin2,
    {
      rules: Object.fromEntries(
        plugin2.ruleDefinitions.filter((rule) => namesOfRulesToEnable.includes(rule.name)).map((rule) => [rule.name, "error"])
      )
    }
  ];
}
function createConfigs(configs2) {
  return configs2;
}
function createPlugin(plugin2) {
  return plugin2;
}

// src/constants.ts
var NAMESPACE = "fsd";

// src/ambiguous-slice-names/index.ts
var ambiguousSliceNames = {
  name: `${NAMESPACE}/ambiguous-slice-names`,
  check(root) {
    const diagnostics = [];
    const layers = getLayers(root);
    const sharedLayer = layers.shared;
    if (sharedLayer === undefined) {
      return { diagnostics };
    }
    const segmentNamesInShared = Object.keys(getSegments(sharedLayer));
    for (const [sliceName, slice] of Object.entries(getAllSlices(root))) {
      const pathSegments = sliceName.split(sep);
      const matchingSegment = segmentNamesInShared.find((segmentName) => pathSegments.includes(segmentName));
      if (matchingSegment === undefined) {
        continue;
      }
      if (matchingSegment === sliceName) {
        diagnostics.push({
          message: `Slice "${sliceName}" could be confused with a segment from Shared with the same name`,
          location: { path: slice.path }
        });
        continue;
      }
      const segmentIndex = pathSegments.indexOf(matchingSegment);
      const isGroup = segmentIndex < pathSegments.length - 1;
      if (!isGroup) {
        diagnostics.push({
          message: `Slice "${sliceName}" could be confused with a segment "${matchingSegment}" from Shared`,
          location: { path: slice.path }
        });
      } else {
        const layer = layers[slice.layerName];
        if (layer === undefined) {
          continue;
        }
        const pathSegmentsToGroup = pathSegments.slice(0, segmentIndex + 1);
        const group = pathSegmentsToGroup.reduce(
          (folder, segmentName) => folder?.children.find((child) => child.type === "folder" && basename(child.path) === segmentName),
          layer
        );
        if (group !== undefined) {
          diagnostics.push({
            message: `Slice group "${pathSegmentsToGroup.join(sep)}" could be confused with a segment "${matchingSegment}" from Shared`,
            location: { path: group.path }
          });
        }
      }
    }
    return { diagnostics };
  }
};
var ambiguous_slice_names_default = ambiguousSliceNames;
function groupSlices(sliceNames) {
  const groups = {};
  for (const sliceName of sliceNames) {
    const segments = sliceName.split(sep);
    const name = segments.pop();
    const group = segments.join(sep);
    if (name !== undefined) {
      groups[group] ??= [];
      groups[group].push(name);
    }
  }
  return groups;
}

// src/excessive-slicing/index.ts
var THRESHOLDS = {
  entities: 20,
  features: 20,
  widgets: 20,
  pages: 20
};
var excessiveSlicing = {
  name: `${NAMESPACE}/excessive-slicing`,
  check(root) {
    const diagnostics = [];
    for (const [layerName, layer] of Object.entries(getLayers(root))) {
      if (!isSliced(layer) || !(layerName in THRESHOLDS)) {
        continue;
      }
      const sliceGroups = groupSlices(Object.keys(getSlices(layer)));
      const threshold = THRESHOLDS[layerName];
      for (const [group, slices] of Object.entries(sliceGroups)) {
        if (slices.length <= threshold) {
          continue;
        }
        if (group === "") {
          diagnostics.push({
            message: `Layer "${layerName}" has ${slices.length} ungrouped slices, which is above the recommended threshold of ${threshold}. Consider grouping them or moving the code inside to the layer where it's used.`,
            location: { path: layer.path }
          });
        } else {
          diagnostics.push({
            message: `Slice group "${group}" has ${slices.length} slices, which is above the recommended threshold of ${threshold}. Consider grouping them or moving the code inside to the layer where it's used.`,
            location: { path: join(layer.path, group) }
          });
        }
      }
    }
    return { diagnostics };
  }
};
var excessive_slicing_default = excessiveSlicing;
function indexSourceFiles(root) {
  const index = {};
  function walk(node, metadata) {
    if (node.type === "file") {
      index[node.path] = { ...metadata, file: node };
    } else {
      for (const child of node.children) {
        walk(child, metadata);
      }
    }
  }
  for (const [layerName, layer] of Object.entries(getLayers(root))) {
    layer.children.filter((child) => child.type === "file").forEach((file) => walk(file, { layerName, sliceName: null, segmentName: null }));
    if (!isSliced(layer)) {
      for (const [segmentName, segment] of Object.entries(getSegments(layer))) {
        walk(segment, { layerName, sliceName: null, segmentName });
      }
    } else {
      for (const [sliceName, slice] of Object.entries(getSlices(layer))) {
        for (const [segmentName, segment] of Object.entries(getSegments(slice))) {
          walk(segment, { layerName, sliceName, segmentName });
        }
        const indexes = getIndexes(slice);
        for (const index2 of indexes) {
          walk(index2, { layerName, sliceName, segmentName: null });
        }
      }
    }
  }
  return index;
}
function collectRelatedTsConfigs(payload) {
  const configs2 = [];
  const setTsConfigs = ({ tsconfig, tsconfigFile, referenced, extended }) => {
    configs2.push(resolveRelativePathsInMergedConfig({ tsconfig, tsconfigFile, extended }));
    referenced?.forEach(setTsConfigs);
  };
  setTsConfigs(payload);
  return configs2;
}
function resolveRelativePathsInMergedConfig(configParseResult) {
  const { tsconfig: mergedConfig, extended } = configParseResult;
  if (
    // if there's no config it needs to be handled somewhere else
    !mergedConfig || // if the merged config does not have compilerOptions, there is nothing to resolve
    !mergedConfig.compilerOptions || // if the merged config does not have paths in compilerOptions there's nothing to resolve
    !mergedConfig.compilerOptions.paths
  ) {
    return mergedConfig;
  }
  const firstConfigWithPaths = findFirstConfigWithPaths(extended || []) || configParseResult;
  const absolutePaths = makeRelativePathAliasesAbsolute(configParseResult, firstConfigWithPaths);
  return {
    ...mergedConfig,
    compilerOptions: {
      ...mergedConfig.compilerOptions,
      paths: absolutePaths
    }
  };
}
function findFirstConfigWithPaths(extended) {
  for (const parseResult of extended) {
    const { tsconfig } = parseResult;
    if (tsconfig.compilerOptions?.paths) {
      return parseResult;
    }
  }
}
function makeRelativePathAliasesAbsolute(finalConfig, firstConfigWithPaths) {
  const { tsconfig: mergedConfig } = finalConfig;
  const {
    compilerOptions: { paths, baseUrl }
  } = mergedConfig;
  const absolutePaths = {};
  if (!firstConfigWithPaths.tsconfigFile) {
    return mergedConfig.compilerOptions.paths;
  }
  for (const entries of Object.entries(paths)) {
    const [key, paths2] = entries;
    absolutePaths[key] = paths2.map(
      (relativePath) => baseUrl ? resolve(dirname(firstConfigWithPaths.tsconfigFile), baseUrl, relativePath) : resolve(dirname(firstConfigWithPaths.tsconfigFile), relativePath)
    );
  }
  return absolutePaths;
}
function resolveDependency(importedPath, importerPath, tsConfigs, fileExists, directoryExists) {
  let resolvedDependency = null;
  for (const config of tsConfigs) {
    resolvedDependency = resolveImport(
      importedPath,
      importerPath,
      config?.compilerOptions ?? {},
      fileExists,
      directoryExists
    );
    if (resolvedDependency !== null) {
      break;
    }
  }
  return resolvedDependency;
}

// src/forbidden-imports/index.ts
var { paperwork } = precinct;
var forbiddenImports = {
  name: `${NAMESPACE}/forbidden-imports`,
  async check(root) {
    const diagnostics = [];
    const parseResult = await parse(root.children[0]?.path ?? root.path);
    const tsConfigs = collectRelatedTsConfigs(parseResult);
    const sourceFileIndex = indexSourceFiles(root);
    for (const sourceFile of Object.values(sourceFileIndex)) {
      const dependencies = paperwork(sourceFile.file.path, { includeCore: false, fileSystem: fs });
      for (const dependency of dependencies) {
        const resolvedDependency = resolveDependency(
          dependency,
          sourceFile.file.path,
          tsConfigs,
          fs.existsSync,
          fs.existsSync
        );
        if (resolvedDependency === null) {
          continue;
        }
        const dependencyLocation = sourceFileIndex[resolvedDependency];
        if (dependencyLocation === undefined) {
          continue;
        }
        if (sourceFile.layerName === dependencyLocation.layerName && sourceFile.sliceName !== dependencyLocation.sliceName) {
          if (dependencyLocation.sliceName !== null && sourceFile.sliceName !== null && !isCrossImportPublicApi(dependencyLocation.file, {
            inSlice: dependencyLocation.sliceName,
            forSlice: sourceFile.sliceName,
            layerPath: join(root.path, dependencyLocation.layerName)
          })) {
            diagnostics.push({
              message: `Forbidden cross-import from slice "${dependencyLocation.sliceName}".`,
              location: { path: sourceFile.file.path }
            });
          }
        } else {
          const thisLayerIndex = layerSequence.indexOf(sourceFile.layerName);
          const dependencyLayerIndex = layerSequence.indexOf(dependencyLocation.layerName);
          if (thisLayerIndex < dependencyLayerIndex) {
            diagnostics.push({
              message: `Forbidden import from higher layer "${dependencyLocation.layerName}".`,
              location: { path: sourceFile.file.path }
            });
          }
        }
      }
    }
    return { diagnostics };
  }
};
var forbidden_imports_default = forbiddenImports;
var { isPlural, plural, singular } = pluralize;
var inconsistentNaming = {
  name: `${NAMESPACE}/inconsistent-naming`,
  check(root) {
    const diagnostics = [];
    const { entities } = getLayers(root);
    if (entities === undefined) {
      return { diagnostics };
    }
    const slices = getSlices(entities);
    const sliceNames = groupSlices(Object.keys(slices));
    for (const [groupPrefix, group] of Object.entries(sliceNames)) {
      const [pluralNames, singularNames] = partition(group, isPlural);
      if (pluralNames.length > 0 && singularNames.length > 0) {
        const message = "Inconsistent pluralization of slice names";
        if (pluralNames.length >= singularNames.length) {
          diagnostics.push({
            message: `${message}. Prefer all plural names`,
            fixes: singularNames.map((name) => ({
              type: "rename",
              path: join(entities.path, groupPrefix, name),
              newName: plural(name)
            })),
            location: { path: join(entities.path, groupPrefix) }
          });
        } else {
          diagnostics.push({
            message: `${message}. Prefer all singular names`,
            fixes: pluralNames.map((name) => ({
              type: "rename",
              path: join(entities.path, groupPrefix, name),
              newName: singular(name)
            })),
            location: { path: join(entities.path, groupPrefix) }
          });
        }
      }
    }
    return { diagnostics };
  }
};
var inconsistent_naming_default = inconsistentNaming;
var { paperwork: paperwork2 } = precinct;
var insignificantSlice = {
  name: `${NAMESPACE}/insignificant-slice`,
  async check(root) {
    const diagnostics = [];
    const references = await traceSliceReferences(root);
    for (const [sourceLocationKey, targetLocationKeys] of references) {
      const [sourceLayerName] = sourceLocationKey.split(sep, 2);
      if (!isSliced(sourceLayerName) || sourceLayerName === "pages") {
        continue;
      }
      if (targetLocationKeys.size === 1) {
        const referenceLocationKey = [...targetLocationKeys][0];
        if (unslicedLayers.includes(referenceLocationKey)) {
          if (referenceLocationKey !== "app") {
            diagnostics.push({
              message: `This slice has only one reference on layer "${referenceLocationKey}". Consider moving this code to "${referenceLocationKey}".`,
              location: { path: join(root.path, sourceLocationKey) }
            });
          }
        } else {
          diagnostics.push({
            message: `This slice has only one reference in slice "${referenceLocationKey}". Consider merging them.`,
            location: { path: join(root.path, sourceLocationKey) }
          });
        }
      } else if (targetLocationKeys.size === 0) {
        diagnostics.push({
          message: `This slice has no references. Consider removing it.`,
          location: { path: join(root.path, sourceLocationKey) }
        });
      }
    }
    return { diagnostics };
  }
};
var insignificant_slice_default = insignificantSlice;
async function traceSliceReferences(root) {
  const sourceFileIndex = indexSourceFiles(root);
  const parseResult = await parse(root.children[0]?.path ?? root.path);
  const tsConfigs = collectRelatedTsConfigs(parseResult);
  const references = /* @__PURE__ */ new Map();
  for (const sourceFile of Object.values(sourceFileIndex)) {
    const thisFileLocationKey = [sourceFile.layerName, sourceFile.sliceName].filter(Boolean).join(sep);
    const dependencies = paperwork2(sourceFile.file.path, { includeCore: false, fileSystem: fs });
    for (const dependency of dependencies) {
      const resolvedDependency = resolveDependency(
        dependency,
        sourceFile.file.path,
        tsConfigs,
        fs.existsSync,
        fs.existsSync
      );
      if (resolvedDependency === null) {
        continue;
      }
      const dependencyLocation = sourceFileIndex[resolvedDependency];
      if (dependencyLocation === undefined || sourceFile.layerName === dependencyLocation.layerName) {
        continue;
      }
      const dependencyLocationKey = [dependencyLocation.layerName, dependencyLocation.sliceName].filter(Boolean).join(sep);
      if (thisFileLocationKey !== dependencyLocationKey) {
        if (!references.has(dependencyLocationKey)) {
          references.set(dependencyLocationKey, /* @__PURE__ */ new Set());
        }
        references.get(dependencyLocationKey)?.add(thisFileLocationKey);
      }
    }
    if (!references.has(thisFileLocationKey)) {
      references.set(thisFileLocationKey, /* @__PURE__ */ new Set());
    }
  }
  return references;
}
var exceptionLayers = ["app"];
var noLayerPublicApi = {
  name: `${NAMESPACE}/no-layer-public-api`,
  check(root) {
    const diagnostics = [];
    for (const [layerName, layer] of Object.entries(getLayers(root))) {
      const indexes = getIndexes(layer);
      const notAmongExceptions = !exceptionLayers.includes(layerName);
      if (notAmongExceptions) {
        for (const index of indexes) {
          diagnostics.push({
            message: `Layer "${layerName}" should not have an index file`,
            location: { path: index.path }
          });
        }
      }
    }
    return { diagnostics };
  }
};
var no_layer_public_api_default = noLayerPublicApi;
var { paperwork: paperwork3 } = precinct;
var noPublicApiSidestep = {
  name: `${NAMESPACE}/no-public-api-sidestep`,
  async check(root) {
    const diagnostics = [];
    const parseResult = await parse(root.children[0]?.path ?? root.path);
    const tsConfigs = collectRelatedTsConfigs(parseResult);
    const sourceFileIndex = indexSourceFiles(root);
    for (const sourceFile of Object.values(sourceFileIndex)) {
      const dependencies = paperwork3(sourceFile.file.path, { includeCore: false, fileSystem: fs });
      for (const dependency of dependencies) {
        const resolvedDependency = resolveDependency(
          dependency,
          sourceFile.file.path,
          tsConfigs,
          fs.existsSync,
          fs.existsSync
        );
        if (resolvedDependency === null) {
          continue;
        }
        const dependencyLocation = sourceFileIndex[resolvedDependency];
        if (dependencyLocation === undefined) {
          continue;
        }
        if (sourceFile.layerName === dependencyLocation.layerName && (!isSliced(dependencyLocation.layerName) || sourceFile.sliceName === dependencyLocation.sliceName)) {
          continue;
        }
        if (isSliced(dependencyLocation.layerName)) {
          if (dependencyLocation.segmentName !== null && dependencyLocation.segmentName !== crossReferenceToken) {
            diagnostics.push({
              message: `Forbidden sidestep of public API when importing from "${dependency}".`,
              location: { path: sourceFile.file.path }
            });
          }
        } else if (dependencyLocation.segmentName !== null) {
          const layer = getLayers(root)[dependencyLocation.layerName];
          if (layer === undefined) {
            continue;
          }
          const segment = getSegments(layer)[dependencyLocation.segmentName];
          if (segment === undefined) {
            continue;
          }
          const segmentIndexes = getIndexes(segment);
          if (segment.type === "folder" && !segmentIndexes.map((segmentIndex) => segmentIndex.path).includes(resolvedDependency)) {
            const { layerName, segmentName } = dependencyLocation;
            if (layerName === "shared" && ["ui", "lib"].includes(segmentName)) {
              const pathInSegment = relative(segment.path, resolvedDependency);
              const topLevelFolder = segment.children.find(
                (child) => child.type === "folder" && basename(child.path) === pathInSegment.split(sep)[0]
              );
              if (topLevelFolder !== undefined) {
                const topLevelFolderIndexes = getIndexes(topLevelFolder);
                if (!topLevelFolderIndexes.map((topLevelFolder2) => topLevelFolder2.path).includes(resolvedDependency)) {
                  diagnostics.push({
                    message: `Forbidden sidestep of public API when importing from "${dependency}".`,
                    location: { path: sourceFile.file.path }
                  });
                }
              }
            } else {
              diagnostics.push({
                message: `Forbidden sidestep of public API when importing from "${dependency}".`,
                location: { path: sourceFile.file.path }
              });
            }
          }
        }
      }
    }
    return { diagnostics };
  }
};
var no_public_api_sidestep_default = noPublicApiSidestep;
var noReservedFolderNames = {
  name: `${NAMESPACE}/no-reserved-folder-names`,
  check(root) {
    const diagnostics = [];
    for (const { segment } of getAllSegments(root)) {
      if (segment.type === "file") {
        continue;
      }
      for (const child of segment.children) {
        if (child.type === "file") {
          continue;
        }
        for (const violatingFolder of findAllRecursively(
          child,
          (entry) => entry.type === "folder" && conventionalSegmentNames.concat(crossReferenceToken).includes(basename(entry.path))
        )) {
          const reservedName = basename(violatingFolder.path);
          if (reservedName === crossReferenceToken) {
            diagnostics.push({
              message: `Having a folder with the name "${crossReferenceToken}" inside a segment could be confusing because that name is reserved for cross-import public APIs. Consider renaming it.`,
              location: { path: violatingFolder.path }
            });
          } else {
            diagnostics.push({
              message: `Having a folder with the name "${basename(violatingFolder.path)}" inside a segment could be confusing because that name is commonly used for segments. Consider renaming it.`,
              location: { path: violatingFolder.path }
            });
          }
        }
      }
    }
    return { diagnostics };
  }
};
var no_reserved_folder_names_default = noReservedFolderNames;
var noSegmentlessSlices = {
  name: `${NAMESPACE}/no-segmentless-slices`,
  check(root) {
    const diagnostics = [];
    for (const layer of Object.values(getLayers(root))) {
      if (!isSliced(layer)) {
        continue;
      }
      let sliceCandidates = layer.children.filter((child) => child.type === "folder");
      while (sliceCandidates.length > 0) {
        const sliceCandidate = sliceCandidates.pop();
        if (sliceCandidate === undefined) {
          continue;
        }
        if (isSliceGroup(sliceCandidate)) {
          sliceCandidates = sliceCandidates.concat(sliceCandidate.children);
        } else {
          if (!isSlice(sliceCandidate)) {
            diagnostics.push({
              message: "This slice has no segments. Consider dividing the code inside into segments.",
              location: { path: sliceCandidate.path }
            });
          }
        }
      }
    }
    return { diagnostics };
  }
};
var no_segmentless_slices_default = noSegmentlessSlices;
function isSliceGroup(folder) {
  return folder.children.length > 0 && !isSlice(folder) && folder.children.every(
    (child) => child.type === "folder" && (isSlice(child) || child.children.every((grandchild) => grandchild.type === "file") || isSliceGroup(child))
  );
}
var noSegmentsOnSlicedLayers = {
  name: `${NAMESPACE}/no-segments-on-sliced-layers`,
  check(root) {
    const diagnostics = [];
    const layers = Object.values(getLayers(root));
    for (const layer of layers) {
      if (isSliced(layer)) {
        for (const directChild of layer.children) {
          if (directChild.type === "folder") {
            const folderName = basename(directChild.path);
            const isConventionalSegment = folderName && conventionalSegmentNames.includes(folderName);
            if (isConventionalSegment) {
              diagnostics.push({
                message: `Conventional segment "${folderName}" should not be a direct child of a sliced layer. Consider moving it inside a slice or, if that is a slice, consider a different name for it to avoid confusion with segments.`,
                location: { path: directChild.path }
              });
            }
          }
        }
      }
    }
    return { diagnostics };
  }
};
var no_segments_on_sliced_layers_default = noSegmentsOnSlicedLayers;
var noUiInApp = {
  name: `${NAMESPACE}/no-ui-in-app`,
  check(root) {
    const diagnostics = [];
    const layers = getLayers(root);
    if (layers.app !== undefined) {
      const segments = getSegments(layers.app);
      if (segments.ui !== undefined) {
        diagnostics.push({
          message: 'Layer "app" should not have "ui" segment.',
          location: { path: segments.ui.path }
        });
      }
    }
    return { diagnostics };
  }
};
var no_ui_in_app_default = noUiInApp;
var publicApi = {
  name: `${NAMESPACE}/public-api`,
  check(root) {
    const diagnostics = [];
    for (const [layerName, layer] of Object.entries(getLayers(root))) {
      if (!isSliced(layer)) {
        if (layerName === "app") {
          continue;
        }
        for (const [segmentName, segment] of Object.entries(getSegments(layer))) {
          if (getIndexes(segment).length === 0) {
            if (!["ui", "lib"].includes(segmentName)) {
              diagnostics.push({
                message: "This segment is missing a public API.",
                fixes: [
                  {
                    type: "create-file",
                    path: join(segment.path, `index.js`),
                    // TODO: Infer better content for this file
                    content: ""
                  }
                ],
                location: { path: segment.path }
              });
            } else if (segment.type === "folder") {
              for (const child of segment.children) {
                if (child.type === "folder" && getIndexes(child).length === 0) {
                  diagnostics.push({
                    message: `This top-level folder in shared/${segmentName} is missing a public API.`,
                    fixes: [
                      {
                        type: "create-file",
                        path: join(child.path, `index.js`),
                        content: ""
                      }
                    ],
                    location: { path: child.path }
                  });
                }
              }
            }
          }
        }
      } else {
        for (const slice of Object.values(getSlices(layer))) {
          if (getIndexes(slice).length === 0) {
            diagnostics.push({
              message: "This slice is missing a public API.",
              fixes: [
                {
                  type: "create-file",
                  path: join(slice.path, `index.js`),
                  // TODO: Infer better content for this file
                  content: ""
                }
              ],
              location: { path: slice.path }
            });
          }
        }
      }
    }
    return { diagnostics };
  }
};
var public_api_default = publicApi;
var wordPattern = /(?:[A-Z]+|[a-z]+)[a-z]*/g;
var repetitiveNaming = {
  name: `${NAMESPACE}/repetitive-naming`,
  check(root) {
    const diagnostics = [];
    for (const layer of Object.values(getLayers(root))) {
      if (!isSliced(layer)) {
        continue;
      }
      const sliceGroups = groupSlices(Object.keys(getSlices(layer)));
      for (const [group, sliceNames] of Object.entries(sliceGroups)) {
        const wordsInSliceNames = sliceNames.map(
          (name) => (name.match(wordPattern) ?? []).map((word) => word.toLowerCase())
        );
        const mostCommonWords = wordsInSliceNames.flat().reduce((acc, word) => {
          acc.set(word, (acc.get(word) ?? 0) + 1);
          return acc;
        }, /* @__PURE__ */ new Map());
        for (const [word, count] of mostCommonWords.entries()) {
          if (sliceNames.length > 2 && count >= sliceNames.length && wordsInSliceNames.every((words) => words.includes(word))) {
            diagnostics.push({
              message: `Repetitive word "${word}" in slice names.`,
              location: { path: join(layer.path, group) }
            });
          }
        }
      }
    }
    return { diagnostics };
  }
};
var repetitive_naming_default = repetitiveNaming;
var BAD_NAMES = ["components", "hooks", "helpers", "utils", "modals", "types", "constants", "consts", "const"];
var segmentsByPurpose = {
  name: `${NAMESPACE}/segments-by-purpose`,
  check(root) {
    const diagnostics = [];
    for (const layer of Object.values(getLayers(root))) {
      if (layer === null) {
        continue;
      }
      if (!isSliced(layer)) {
        for (const [segmentName, segment] of Object.entries(getSegments(layer))) {
          if (BAD_NAMES.includes(segmentName)) {
            diagnostics.push({
              message: "This segment's name should describe the purpose of its contents, not what the contents are.",
              location: { path: segment.path }
            });
          }
        }
      } else {
        for (const slice of Object.values(getSlices(layer))) {
          for (const [segmentName, segment] of Object.entries(getSegments(slice))) {
            if (BAD_NAMES.includes(segmentName)) {
              diagnostics.push({
                message: "This segment's name should describe the purpose of its contents, not what the contents are.",
                location: { path: segment.path }
              });
            }
          }
        }
      }
    }
    return { diagnostics };
  }
};
var segments_by_purpose_default = segmentsByPurpose;
var THRESHOLD = 15;
var sharedLibGrouping = {
  name: `${NAMESPACE}/shared-lib-grouping`,
  check(root) {
    const diagnostics = [];
    const { shared } = getLayers(root);
    if (!shared) {
      return { diagnostics };
    }
    const { lib } = getSegments(shared);
    if (!lib) {
      return { diagnostics };
    }
    if (lib.type === "folder" && lib.children.length > THRESHOLD) {
      diagnostics.push({
        message: `Shared/lib has ${lib.children.length} modules, which is above the recommended threshold of ${THRESHOLD}. Consider grouping them.`,
        location: { path: lib.path }
      });
    }
    return { diagnostics };
  }
};
var shared_lib_grouping_default = sharedLibGrouping;
var LEVENSHTEIN_DISTANCE_UPPER_BOUND = 3;
var typoInLayerName = {
  name: `${NAMESPACE}/typo-in-layer-name`,
  check(root) {
    const diagnostics = [];
    const suggestionsList = root.children.filter((child) => child.type === "folder").flatMap((child) => {
      const layer = basename(child.path);
      return layerSequence.map((sequenceLayer) => ({
        input: layer,
        suggestion: sequenceLayer,
        distance: distance(layer, sequenceLayer)
      })).filter((layer2) => layer2.distance <= LEVENSHTEIN_DISTANCE_UPPER_BOUND);
    }).sort((a, b) => a.distance - b.distance);
    const processedInputs = [];
    const suggestedLayers = [];
    suggestionsList.forEach((layer) => {
      if (layer.distance === 0) {
        suggestedLayers.push(layer.suggestion);
        return;
      }
      if (processedInputs.includes(layer.input)) {
        return;
      }
      if (suggestedLayers.includes(layer.suggestion)) {
        return;
      }
      processedInputs.push(layer.input);
      suggestedLayers.push(layer.suggestion);
      diagnostics.push({
        message: `Layer "${layer.input}" potentially contains a typo. Did you mean "${layer.suggestion}"?`,
        location: { path: join(root.path, layer.input) }
      });
    });
    return { diagnostics };
  }
};
var typo_in_layer_name_default = typoInLayerName;
var noProcesses = {
  name: `${NAMESPACE}/no-processes`,
  check(root) {
    const diagnostics = [];
    const processesLayer = root.children.find(
      (child) => child.type === "folder" && basename(child.path) === "processes"
    );
    if (processesLayer !== undefined) {
      diagnostics.push({
        message: 'Layer "processes" is deprecated, avoid using it',
        location: { path: processesLayer.path }
      });
    }
    return { diagnostics };
  }
};
var no_processes_default = noProcesses;

// package.json
var package_default = {
  name: "@feature-sliced/steiger-plugin",
  description: "A set of rules to verify correct usage of Feature-Sliced Design",
  version: "0.5.7",
  scripts: {
    lint: "eslint .",
    format: "prettier --write . --cache",
    "check-formatting": "prettier --check . --cache",
    dev: "tsup --watch",
    build: "tsup",
    test: "vitest run",
    typecheck: "tsc --noEmit"
  },
  exports: {
    types: "./dist/index.d.ts",
    import: "./dist/index.js"
  },
  files: [
    "dist",
    "src"
  ],
  type: "module",
  license: "MIT",
  authors: [
    {
      name: "Anton Medvedev",
      email: "unordinarity@yandex.ru",
      url: "https://github.com/unordinarity"
    },
    {
      name: "Lev Chelyadinov",
      email: "leva181777@gmail.com",
      url: "https://github.com/illright"
    }
  ],
  homepage: "https://github.com/feature-sliced/steiger",
  repository: {
    type: "git",
    url: "git+https://github.com/feature-sliced/steiger.git",
    directory: "packages/steiger-plugin-fsd"
  },
  dependencies: {
    "@feature-sliced/filesystem": "^3.0.1",
    "fastest-levenshtein": "^1.0.16",
    "lodash-es": "^4.17.21",
    pluralize: "^8.0.0",
    precinct: "^12.1.2",
    tsconfck: "^3.1.4"
  },
  devDependencies: {
    "@steiger/eslint-config": "workspace:*",
    "@steiger/toolkit": "workspace:*",
    "@steiger/tsconfig": "workspace:*",
    "@total-typescript/ts-reset": "^0.6.1",
    "@types/lodash-es": "^4.17.12",
    "@types/pluralize": "^0.0.33",
    tsup: "^8.3.5",
    typescript: "^5.7.3",
    vitest: "^3.0.2"
  }
};
var { paperwork: paperwork4 } = precinct;
var noCrossImports = {
  name: `${NAMESPACE}/no-cross-imports`,
  async check(root) {
    const diagnostics = [];
    const parseResult = await parse(root.children[0]?.path ?? root.path);
    const tsConfigs = collectRelatedTsConfigs(parseResult);
    const sourceFileIndex = indexSourceFiles(root);
    for (const sourceFile of Object.values(sourceFileIndex)) {
      const dependencies = paperwork4(sourceFile.file.path, { includeCore: false, fileSystem: fs });
      for (const dependency of dependencies) {
        const resolvedDependency = resolveDependency(
          dependency,
          sourceFile.file.path,
          tsConfigs,
          fs.existsSync,
          fs.existsSync
        );
        if (resolvedDependency === null) {
          continue;
        }
        const dependencyLocation = sourceFileIndex[resolvedDependency];
        if (dependencyLocation === undefined) {
          continue;
        }
        if (sourceFile.layerName === dependencyLocation.layerName && sourceFile.sliceName !== dependencyLocation.sliceName) {
          if (dependencyLocation.sliceName !== null && sourceFile.sliceName !== null && !isCrossImportPublicApi(dependencyLocation.file, {
            inSlice: dependencyLocation.sliceName,
            forSlice: sourceFile.sliceName,
            layerPath: join(root.path, dependencyLocation.layerName)
          })) {
            diagnostics.push({
              message: `Forbidden cross-import from slice "${dependencyLocation.sliceName}".`,
              location: { path: sourceFile.file.path }
            });
          }
        }
      }
    }
    return { diagnostics };
  }
};
var no_cross_imports_default = noCrossImports;
var { paperwork: paperwork5 } = precinct;
var noHigherLevelImports = {
  name: `${NAMESPACE}/no-higher-level-imports`,
  async check(root) {
    const diagnostics = [];
    const parseResult = await parse(root.children[0]?.path ?? root.path);
    const tsConfigs = collectRelatedTsConfigs(parseResult);
    const sourceFileIndex = indexSourceFiles(root);
    for (const sourceFile of Object.values(sourceFileIndex)) {
      const dependencies = paperwork5(sourceFile.file.path, { includeCore: false, fileSystem: fs });
      for (const dependency of dependencies) {
        const resolvedDependency = resolveDependency(
          dependency,
          sourceFile.file.path,
          tsConfigs,
          fs.existsSync,
          fs.existsSync
        );
        if (resolvedDependency === null) {
          continue;
        }
        const dependencyLocation = sourceFileIndex[resolvedDependency];
        if (dependencyLocation === undefined) {
          continue;
        }
        const thisLayerIndex = layerSequence.indexOf(sourceFile.layerName);
        const dependencyLayerIndex = layerSequence.indexOf(dependencyLocation.layerName);
        if (thisLayerIndex < dependencyLayerIndex) {
          diagnostics.push({
            message: `Forbidden import from higher layer "${dependencyLocation.layerName}".`,
            location: { path: sourceFile.file.path }
          });
        }
      }
    }
    return { diagnostics };
  }
};
var no_higher_level_imports_default = noHigherLevelImports;

// src/index.ts
var enabledRules = [
  ambiguous_slice_names_default,
  excessive_slicing_default,
  forbidden_imports_default,
  inconsistent_naming_default,
  insignificant_slice_default,
  no_layer_public_api_default,
  no_public_api_sidestep_default,
  no_reserved_folder_names_default,
  no_segmentless_slices_default,
  no_segments_on_sliced_layers_default,
  no_ui_in_app_default,
  public_api_default,
  repetitive_naming_default,
  segments_by_purpose_default,
  shared_lib_grouping_default,
  typo_in_layer_name_default,
  no_processes_default
];
var disabledRules = [no_cross_imports_default, no_higher_level_imports_default];
var rules = [...enabledRules, ...disabledRules];
var plugin = createPlugin({
  meta: {
    name: "@feature-sliced/steiger-plugin",
    version: package_default.version
  },
  ruleDefinitions: rules
});
var configs = createConfigs({
  recommended: enableSpecificRules(plugin, enabledRules)
});
var index_default = {
  plugin,
  configs
};

export { index_default as default };
