// @gfazioli/mantine-spinner/dist/cjs/Spinner.cjs
'use client';
'use strict';

var React = require('react');
var core = require('@mantine/core');
var Spinner_module = require('./Spinner.module.css.cjs');

const defaultProps = {
  size: "md",
  inner: 8,
  segments: 12,
  thickness: 3,
  speed: 1200,
  direction: "clockwise",
  strokeLinecap: "round",
  transitionTimingFunction: "ease"
};
const varsResolver = core.createVarsResolver(
  (_, { strokeLinecap, speed, transitionTimingFunction }) => {
    return {
      root: {
        "--spinner-stroke-linecap": strokeLinecap,
        "--spinner-animation-duration": `${speed || 1}ms`,
        "--spinner-timing-function": transitionTimingFunction
      }
    };
  }
);
const Spinner = core.factory((_props, ref) => {
  const [isClient, setIsClient] = React.useState(false);
  const props = core.useProps("Spinner", defaultProps, _props);
  const theme = core.useMantineTheme();
  const [invalidate, setInvalidate] = React.useState(false);
  React.useEffect(() => {
    setIsClient(true);
  }, []);
  const {
    size,
    inner,
    segments,
    thickness,
    speed,
    color,
    direction,
    transitionTimingFunction,
    strokeLinecap,
    classNames,
    style,
    styles,
    unstyled,
    vars,
    className,
    ...others
  } = props;
  const getStyles = core.useStyles({
    name: "Spinner",
    props,
    classes: Spinner_module,
    className,
    style,
    classNames,
    styles,
    unstyled,
    vars,
    varsResolver
  });
  const getSizeValue = (size2) => {
    if (typeof size2 === "string" && ["lg", "xl", "md", "sm", "xs"].includes(size2)) {
      return {
        xl: 58,
        lg: 44,
        md: 36,
        sm: 22,
        xs: 18
      }[size2];
    }
    return core.px(size2);
  };
  const sizeValue = getSizeValue(size);
  const innerValue = getSizeValue(inner);
  const center = sizeValue / 2;
  const maxRadius = center - thickness;
  const radius = Math.min(sizeValue / 2, maxRadius);
  const innerRadius = Math.min(innerValue, radius);
  const directionValue = direction === "counter-clockwise" ? -1 : 1;
  const parsedColor = core.parseThemeColor({
    color: color || theme.primaryColor,
    theme
  });
  React.useEffect(() => {
    if (!invalidate) {
      setInvalidate(true);
      setTimeout(() => setInvalidate(false), 500);
    }
  }, [segments]);
  if (invalidate || !isClient) {
    return null;
  }
  return /* @__PURE__ */ React.createElement(core.Box, { ref }, /* @__PURE__ */ React.createElement(
    core.Box,
    {
      ...getStyles("root"),
      component: "svg",
      ...others,
      xmlns: "http://www.w3.org/2000/svg",
      width: sizeValue,
      height: sizeValue,
      viewBox: `0 0 ${sizeValue} ${sizeValue}`,
      preserveAspectRatio: "xMidYMid meet"
    },
    Array.from({ length: segments }).map((_, index) => {
      const angle = 360 / segments * index - 90;
      const rad = angle * Math.PI / 180;
      const x1 = center + innerRadius * Math.cos(rad);
      const y1 = center + innerRadius * Math.sin(rad);
      const x2 = center + radius * Math.cos(rad);
      const y2 = center + radius * Math.sin(rad);
      return /* @__PURE__ */ React.createElement(
        "line",
        {
          key: `line-${index}`,
          className: Spinner_module.line,
          x1,
          y1,
          x2,
          y2,
          stroke: parsedColor.value,
          strokeWidth: thickness,
          style: {
            animationDelay: `${index * speed * directionValue / segments}ms`
          }
        }
      );
    })
  ));
});
Spinner.classes = Spinner_module;
Spinner.displayName = "Spinner";

exports.Spinner = Spinner;
exports.defaultProps = defaultProps;
